// speed-components-ui@0.1.7/dist/useAntdCssVars-Bzx55LlW.mjs
import { ref as Q, inject as xt, computed as Oe, unref as Fi, reactive as Yi, watch as dn } from "vue";
import { message as it, theme as vn } from "ant-design-vue";
var Mt = typeof globalThis < "u" ? globalThis : typeof window < "u" ? window : typeof global < "u" ? global : typeof self < "u" ? self : {};
function ji(t) {
  return t && t.__esModule && Object.prototype.hasOwnProperty.call(t, "default") ? t.default : t;
}
var gt = { exports: {} }, kt;
function ki() {
  return kt || (kt = 1, function(t, e) {
    (function(i, n) {
      t.exports = n();
    })(Mt, function() {
      var i = 1e3, n = 6e4, r = 36e5, a = "millisecond", s = "second", u = "minute", o = "hour", f = "day", c = "week", l = "month", d = "quarter", b = "year", h = "date", E = "Invalid Date", g = /^(\d{4})[-/]?(\d{1,2})?[-/]?(\d{0,2})[Tt\s]*(\d{1,2})?:?(\d{1,2})?:?(\d{1,2})?[.:]?(\d+)?$/, m = /\[([^\]]+)]|Y{1,4}|M{1,4}|D{1,2}|d{1,4}|H{1,2}|h{1,2}|a|A|m{1,2}|s{1,2}|Z{1,2}|SSS/g, _ = { name: "en", weekdays: "Sunday_Monday_Tuesday_Wednesday_Thursday_Friday_Saturday".split("_"), months: "January_February_March_April_May_June_July_August_September_October_November_December".split("_"), ordinal: function(T) {
        var v = ["th", "st", "nd", "rd"], w = T % 100;
        return "[" + T + (v[(w - 20) % 10] || v[w] || v[0]) + "]";
      } }, S = function(T, v, w) {
        var D = String(T);
        return !D || D.length >= v ? T : "" + Array(v + 1 - D.length).join(w) + T;
      }, P = { s: S, z: function(T) {
        var v = -T.utcOffset(), w = Math.abs(v), D = Math.floor(w / 60), y = w % 60;
        return (v <= 0 ? "+" : "-") + S(D, 2, "0") + ":" + S(y, 2, "0");
      }, m: function T(v, w) {
        if (v.date() < w.date()) return -T(w, v);
        var D = 12 * (w.year() - v.year()) + (w.month() - v.month()), y = v.clone().add(D, l), O = w - y < 0, x = v.clone().add(D + (O ? -1 : 1), l);
        return +(-(D + (w - y) / (O ? y - x : x - y)) || 0);
      }, a: function(T) {
        return T < 0 ? Math.ceil(T) || 0 : Math.floor(T);
      }, p: function(T) {
        return { M: l, y: b, w: c, d: f, D: h, h: o, m: u, s, ms: a, Q: d }[T] || String(T || "").toLowerCase().replace(/s$/, "");
      }, u: function(T) {
        return T === void 0;
      } }, N = "en", F = {};
      F[N] = _;
      var X = "$isDayjsObject", W = function(T) {
        return T instanceof I || !(!T || !T[X]);
      }, R = function T(v, w, D) {
        var y;
        if (!v) return N;
        if (typeof v == "string") {
          var O = v.toLowerCase();
          F[O] && (y = O), w && (F[O] = w, y = O);
          var x = v.split("-");
          if (!y && x.length > 1) return T(x[0]);
        } else {
          var $ = v.name;
          F[$] = v, y = $;
        }
        return !D && y && (N = y), y || !D && N;
      }, A = function(T, v) {
        if (W(T)) return T.clone();
        var w = typeof v == "object" ? v : {};
        return w.date = T, w.args = arguments, new I(w);
      }, p = P;
      p.l = R, p.i = W, p.w = function(T, v) {
        return A(T, { locale: v.$L, utc: v.$u, x: v.$x, $offset: v.$offset });
      };
      var I = function() {
        function T(w) {
          this.$L = R(w.locale, null, !0), this.parse(w), this.$x = this.$x || w.x || {}, this[X] = !0;
        }
        var v = T.prototype;
        return v.parse = function(w) {
          this.$d = function(D) {
            var y = D.date, O = D.utc;
            if (y === null) return /* @__PURE__ */ new Date(NaN);
            if (p.u(y)) return /* @__PURE__ */ new Date();
            if (y instanceof Date) return new Date(y);
            if (typeof y == "string" && !/Z$/i.test(y)) {
              var x = y.match(g);
              if (x) {
                var $ = x[2] - 1 || 0, U = (x[7] || "0").substring(0, 3);
                return O ? new Date(Date.UTC(x[1], $, x[3] || 1, x[4] || 0, x[5] || 0, x[6] || 0, U)) : new Date(x[1], $, x[3] || 1, x[4] || 0, x[5] || 0, x[6] || 0, U);
              }
            }
            return new Date(y);
          }(w), this.init();
        }, v.init = function() {
          var w = this.$d;
          this.$y = w.getFullYear(), this.$M = w.getMonth(), this.$D = w.getDate(), this.$W = w.getDay(), this.$H = w.getHours(), this.$m = w.getMinutes(), this.$s = w.getSeconds(), this.$ms = w.getMilliseconds();
        }, v.$utils = function() {
          return p;
        }, v.isValid = function() {
          return this.$d.toString() !== E;
        }, v.isSame = function(w, D) {
          var y = A(w);
          return this.startOf(D) <= y && y <= this.endOf(D);
        }, v.isAfter = function(w, D) {
          return A(w) < this.startOf(D);
        }, v.isBefore = function(w, D) {
          return this.endOf(D) < A(w);
        }, v.$g = function(w, D, y) {
          return p.u(w) ? this[D] : this.set(y, w);
        }, v.unix = function() {
          return Math.floor(this.valueOf() / 1e3);
        }, v.valueOf = function() {
          return this.$d.getTime();
        }, v.startOf = function(w, D) {
          var y = this, O = !!p.u(D) || D, x = p.p(w), $ = function(oe, J) {
            var ie = p.w(y.$u ? Date.UTC(y.$y, J, oe) : new Date(y.$y, J, oe), y);
            return O ? ie : ie.endOf(f);
          }, U = function(oe, J) {
            return p.w(y.toDate()[oe].apply(y.toDate("s"), (O ? [0, 0, 0, 0] : [23, 59, 59, 999]).slice(J)), y);
          }, q = this.$W, Z = this.$M, ee = this.$D, me = "set" + (this.$u ? "UTC" : "");
          switch (x) {
            case b:
              return O ? $(1, 0) : $(31, 11);
            case l:
              return O ? $(1, Z) : $(0, Z + 1);
            case c:
              var le = this.$locale().weekStart || 0, be = (q < le ? q + 7 : q) - le;
              return $(O ? ee - be : ee + (6 - be), Z);
            case f:
            case h:
              return U(me + "Hours", 0);
            case o:
              return U(me + "Minutes", 1);
            case u:
              return U(me + "Seconds", 2);
            case s:
              return U(me + "Milliseconds", 3);
            default:
              return this.clone();
          }
        }, v.endOf = function(w) {
          return this.startOf(w, !1);
        }, v.$set = function(w, D) {
          var y, O = p.p(w), x = "set" + (this.$u ? "UTC" : ""), $ = (y = {}, y[f] = x + "Date", y[h] = x + "Date", y[l] = x + "Month", y[b] = x + "FullYear", y[o] = x + "Hours", y[u] = x + "Minutes", y[s] = x + "Seconds", y[a] = x + "Milliseconds", y)[O], U = O === f ? this.$D + (D - this.$W) : D;
          if (O === l || O === b) {
            var q = this.clone().set(h, 1);
            q.$d[$](U), q.init(), this.$d = q.set(h, Math.min(this.$D, q.daysInMonth())).$d;
          } else $ && this.$d[$](U);
          return this.init(), this;
        }, v.set = function(w, D) {
          return this.clone().$set(w, D);
        }, v.get = function(w) {
          return this[p.p(w)]();
        }, v.add = function(w, D) {
          var y, O = this;
          w = Number(w);
          var x = p.p(D), $ = function(Z) {
            var ee = A(O);
            return p.w(ee.date(ee.date() + Math.round(Z * w)), O);
          };
          if (x === l) return this.set(l, this.$M + w);
          if (x === b) return this.set(b, this.$y + w);
          if (x === f) return $(1);
          if (x === c) return $(7);
          var U = (y = {}, y[u] = n, y[o] = r, y[s] = i, y)[x] || 1, q = this.$d.getTime() + w * U;
          return p.w(q, this);
        }, v.subtract = function(w, D) {
          return this.add(-1 * w, D);
        }, v.format = function(w) {
          var D = this, y = this.$locale();
          if (!this.isValid()) return y.invalidDate || E;
          var O = w || "YYYY-MM-DDTHH:mm:ssZ", x = p.z(this), $ = this.$H, U = this.$m, q = this.$M, Z = y.weekdays, ee = y.months, me = y.meridiem, le = function(J, ie, ue, pe) {
            return J && (J[ie] || J(D, O)) || ue[ie].slice(0, pe);
          }, be = function(J) {
            return p.s($ % 12 || 12, J, "0");
          }, oe = me || function(J, ie, ue) {
            var pe = J < 12 ? "AM" : "PM";
            return ue ? pe.toLowerCase() : pe;
          };
          return O.replace(m, function(J, ie) {
            return ie || function(ue) {
              switch (ue) {
                case "YY":
                  return String(D.$y).slice(-2);
                case "YYYY":
                  return p.s(D.$y, 4, "0");
                case "M":
                  return q + 1;
                case "MM":
                  return p.s(q + 1, 2, "0");
                case "MMM":
                  return le(y.monthsShort, q, ee, 3);
                case "MMMM":
                  return le(ee, q);
                case "D":
                  return D.$D;
                case "DD":
                  return p.s(D.$D, 2, "0");
                case "d":
                  return String(D.$W);
                case "dd":
                  return le(y.weekdaysMin, D.$W, Z, 2);
                case "ddd":
                  return le(y.weekdaysShort, D.$W, Z, 3);
                case "dddd":
                  return Z[D.$W];
                case "H":
                  return String($);
                case "HH":
                  return p.s($, 2, "0");
                case "h":
                  return be(1);
                case "hh":
                  return be(2);
                case "a":
                  return oe($, U, !0);
                case "A":
                  return oe($, U, !1);
                case "m":
                  return String(U);
                case "mm":
                  return p.s(U, 2, "0");
                case "s":
                  return String(D.$s);
                case "ss":
                  return p.s(D.$s, 2, "0");
                case "SSS":
                  return p.s(D.$ms, 3, "0");
                case "Z":
                  return x;
              }
              return null;
            }(J) || x.replace(":", "");
          });
        }, v.utcOffset = function() {
          return 15 * -Math.round(this.$d.getTimezoneOffset() / 15);
        }, v.diff = function(w, D, y) {
          var O, x = this, $ = p.p(D), U = A(w), q = (U.utcOffset() - this.utcOffset()) * n, Z = this - U, ee = function() {
            return p.m(x, U);
          };
          switch ($) {
            case b:
              O = ee() / 12;
              break;
            case l:
              O = ee();
              break;
            case d:
              O = ee() / 3;
              break;
            case c:
              O = (Z - q) / 6048e5;
              break;
            case f:
              O = (Z - q) / 864e5;
              break;
            case o:
              O = Z / r;
              break;
            case u:
              O = Z / n;
              break;
            case s:
              O = Z / i;
              break;
            default:
              O = Z;
          }
          return y ? O : p.a(O);
        }, v.daysInMonth = function() {
          return this.endOf(l).$D;
        }, v.$locale = function() {
          return F[this.$L];
        }, v.locale = function(w, D) {
          if (!w) return this.$L;
          var y = this.clone(), O = R(w, D, !0);
          return O && (y.$L = O), y;
        }, v.clone = function() {
          return p.w(this.$d, this);
        }, v.toDate = function() {
          return new Date(this.valueOf());
        }, v.toJSON = function() {
          return this.isValid() ? this.toISOString() : null;
        }, v.toISOString = function() {
          return this.$d.toISOString();
        }, v.toString = function() {
          return this.$d.toUTCString();
        }, T;
      }(), k = I.prototype;
      return A.prototype = k, [["$ms", a], ["$s", s], ["$m", u], ["$H", o], ["$W", f], ["$M", l], ["$y", b], ["$D", h]].forEach(function(T) {
        k[T[1]] = function(v) {
          return this.$g(v, T[0], T[1]);
        };
      }), A.extend = function(T, v) {
        return T.$i || (T(v, I, A), T.$i = !0), A;
      }, A.locale = R, A.isDayjs = W, A.unix = function(T) {
        return A(1e3 * T);
      }, A.en = F[N], A.Ls = F, A.p = {}, A;
    });
  }(gt)), gt.exports;
}
var gn = ki();
const mn = /* @__PURE__ */ ji(gn);
var Hi = { exports: {} };
(function(t, e) {
  (function(i, n) {
    t.exports = n();
  })(Mt, function() {
    var i = { LTS: "h:mm:ss A", LT: "h:mm A", L: "MM/DD/YYYY", LL: "MMMM D, YYYY", LLL: "MMMM D, YYYY h:mm A", LLLL: "dddd, MMMM D, YYYY h:mm A" }, n = /(\[[^[]*\])|([-_:/.,()\s]+)|(A|a|Q|YYYY|YY?|ww?|MM?M?M?|Do|DD?|hh?|HH?|mm?|ss?|S{1,3}|z|ZZ?)/g, r = /\d/, a = /\d\d/, s = /\d\d?/, u = /\d*[^-_:/,()\s\d]+/, o = {}, f = function(g) {
      return (g = +g) + (g > 68 ? 1900 : 2e3);
    }, c = function(g) {
      return function(m) {
        this[g] = +m;
      };
    }, l = [/[+-]\d\d:?(\d\d)?|Z/, function(g) {
      (this.zone || (this.zone = {})).offset = function(m) {
        if (!m || m === "Z") return 0;
        var _ = m.match(/([+-]|\d\d)/g), S = 60 * _[1] + (+_[2] || 0);
        return S === 0 ? 0 : _[0] === "+" ? -S : S;
      }(g);
    }], d = function(g) {
      var m = o[g];
      return m && (m.indexOf ? m : m.s.concat(m.f));
    }, b = function(g, m) {
      var _, S = o.meridiem;
      if (S) {
        for (var P = 1; P <= 24; P += 1) if (g.indexOf(S(P, 0, m)) > -1) {
          _ = P > 12;
          break;
        }
      } else _ = g === (m ? "pm" : "PM");
      return _;
    }, h = { A: [u, function(g) {
      this.afternoon = b(g, !1);
    }], a: [u, function(g) {
      this.afternoon = b(g, !0);
    }], Q: [r, function(g) {
      this.month = 3 * (g - 1) + 1;
    }], S: [r, function(g) {
      this.milliseconds = 100 * +g;
    }], SS: [a, function(g) {
      this.milliseconds = 10 * +g;
    }], SSS: [/\d{3}/, function(g) {
      this.milliseconds = +g;
    }], s: [s, c("seconds")], ss: [s, c("seconds")], m: [s, c("minutes")], mm: [s, c("minutes")], H: [s, c("hours")], h: [s, c("hours")], HH: [s, c("hours")], hh: [s, c("hours")], D: [s, c("day")], DD: [a, c("day")], Do: [u, function(g) {
      var m = o.ordinal, _ = g.match(/\d+/);
      if (this.day = _[0], m) for (var S = 1; S <= 31; S += 1) m(S).replace(/\[|\]/g, "") === g && (this.day = S);
    }], w: [s, c("week")], ww: [a, c("week")], M: [s, c("month")], MM: [a, c("month")], MMM: [u, function(g) {
      var m = d("months"), _ = (d("monthsShort") || m.map(function(S) {
        return S.slice(0, 3);
      })).indexOf(g) + 1;
      if (_ < 1) throw new Error();
      this.month = _ % 12 || _;
    }], MMMM: [u, function(g) {
      var m = d("months").indexOf(g) + 1;
      if (m < 1) throw new Error();
      this.month = m % 12 || m;
    }], Y: [/[+-]?\d+/, c("year")], YY: [a, function(g) {
      this.year = f(g);
    }], YYYY: [/\d{4}/, c("year")], Z: l, ZZ: l };
    function E(g) {
      var m, _;
      m = g, _ = o && o.formats;
      for (var S = (g = m.replace(/(\[[^\]]+])|(LTS?|l{1,4}|L{1,4})/g, function(A, p, I) {
        var k = I && I.toUpperCase();
        return p || _[I] || i[I] || _[k].replace(/(\[[^\]]+])|(MMMM|MM|DD|dddd)/g, function(T, v, w) {
          return v || w.slice(1);
        });
      })).match(n), P = S.length, N = 0; N < P; N += 1) {
        var F = S[N], X = h[F], W = X && X[0], R = X && X[1];
        S[N] = R ? { regex: W, parser: R } : F.replace(/^\[|\]$/g, "");
      }
      return function(A) {
        for (var p = {}, I = 0, k = 0; I < P; I += 1) {
          var T = S[I];
          if (typeof T == "string") k += T.length;
          else {
            var v = T.regex, w = T.parser, D = A.slice(k), y = v.exec(D)[0];
            w.call(p, y), A = A.replace(y, "");
          }
        }
        return function(O) {
          var x = O.afternoon;
          if (x !== void 0) {
            var $ = O.hours;
            x ? $ < 12 && (O.hours += 12) : $ === 12 && (O.hours = 0), delete O.afternoon;
          }
        }(p), p;
      };
    }
    return function(g, m, _) {
      _.p.customParseFormat = !0, g && g.parseTwoDigitYear && (f = g.parseTwoDigitYear);
      var S = m.prototype, P = S.parse;
      S.parse = function(N) {
        var F = N.date, X = N.utc, W = N.args;
        this.$u = X;
        var R = W[1];
        if (typeof R == "string") {
          var A = W[2] === !0, p = W[3] === !0, I = A || p, k = W[2];
          p && (k = W[2]), o = this.$locale(), !A && k && (o = _.Ls[k]), this.$d = function(D, y, O, x) {
            try {
              if (["x", "X"].indexOf(y) > -1) return new Date((y === "X" ? 1e3 : 1) * D);
              var $ = E(y)(D), U = $.year, q = $.month, Z = $.day, ee = $.hours, me = $.minutes, le = $.seconds, be = $.milliseconds, oe = $.zone, J = $.week, ie = /* @__PURE__ */ new Date(), ue = Z || (U || q ? 1 : ie.getDate()), pe = U || ie.getFullYear(), Ke = 0;
              U && !q || (Ke = q > 0 ? q - 1 : ie.getMonth());
              var Ge, ft = ee || 0, ht = me || 0, dt = le || 0, vt = be || 0;
              return oe ? new Date(Date.UTC(pe, Ke, ue, ft, ht, dt, vt + 60 * oe.offset * 1e3)) : O ? new Date(Date.UTC(pe, Ke, ue, ft, ht, dt, vt)) : (Ge = new Date(pe, Ke, ue, ft, ht, dt, vt), J && (Ge = x(Ge).week(J).toDate()), Ge);
            } catch {
              return /* @__PURE__ */ new Date("");
            }
          }(F, R, X, _), this.init(), k && k !== !0 && (this.$L = this.locale(k).$L), I && F != this.format(R) && (this.$d = /* @__PURE__ */ new Date("")), o = {};
        } else if (R instanceof Array) for (var T = R.length, v = 1; v <= T; v += 1) {
          W[1] = R[v - 1];
          var w = _.apply(this, W);
          if (w.isValid()) {
            this.$d = w.$d, this.$L = w.$L, this.init();
            break;
          }
          v === T && (this.$d = /* @__PURE__ */ new Date(""));
        }
        else P.call(this, N);
      };
    };
  });
})(Hi);
var pn = Hi.exports;
const wn = /* @__PURE__ */ ji(pn);
var yn = { exports: {} };
(function(t, e) {
  (function(i, n) {
    t.exports = n(ki());
  })(Mt, function(i) {
    function n(s) {
      return s && typeof s == "object" && "default" in s ? s : { default: s };
    }
    var r = n(i), a = { name: "zh-cn", weekdays: "星期日_星期一_星期二_星期三_星期四_星期五_星期六".split("_"), weekdaysShort: "周日_周一_周二_周三_周四_周五_周六".split("_"), weekdaysMin: "日_一_二_三_四_五_六".split("_"), months: "一月_二月_三月_四月_五月_六月_七月_八月_九月_十月_十一月_十二月".split("_"), monthsShort: "1月_2月_3月_4月_5月_6月_7月_8月_9月_10月_11月_12月".split("_"), ordinal: function(s, u) {
      return u === "W" ? s + "周" : s + "日";
    }, weekStart: 1, yearStart: 4, formats: { LT: "HH:mm", LTS: "HH:mm:ss", L: "YYYY/MM/DD", LL: "YYYY年M月D日", LLL: "YYYY年M月D日Ah点mm分", LLLL: "YYYY年M月D日ddddAh点mm分", l: "YYYY/M/D", ll: "YYYY年M月D日", lll: "YYYY年M月D日 HH:mm", llll: "YYYY年M月D日dddd HH:mm" }, relativeTime: { future: "%s内", past: "%s前", s: "几秒", m: "1 分钟", mm: "%d 分钟", h: "1 小时", hh: "%d 小时", d: "1 天", dd: "%d 天", M: "1 个月", MM: "%d 个月", y: "1 年", yy: "%d 年" }, meridiem: function(s, u) {
      var o = 100 * s + u;
      return o < 600 ? "凌晨" : o < 900 ? "早上" : o < 1100 ? "上午" : o < 1300 ? "中午" : o < 1800 ? "下午" : "晚上";
    } };
    return r.default.locale(a, null, !0), a;
  });
})(yn);
var Ui = typeof global == "object" && global && global.Object === Object && global, bn = typeof self == "object" && self && self.Object === Object && self, de = Ui || bn || Function("return this")(), ye = de.Symbol, Vi = Object.prototype, _n = Vi.hasOwnProperty, Sn = Vi.toString, Ye = ye ? ye.toStringTag : void 0;
function En(t) {
  var e = _n.call(t, Ye), i = t[Ye];
  try {
    t[Ye] = void 0;
    var n = !0;
  } catch {
  }
  var r = Sn.call(t);
  return n && (e ? t[Ye] = i : delete t[Ye]), r;
}
var Tn = Object.prototype, Dn = Tn.toString;
function An(t) {
  return Dn.call(t);
}
var On = "[object Null]", xn = "[object Undefined]", Ht = ye ? ye.toStringTag : void 0;
function ze(t) {
  return t == null ? t === void 0 ? xn : On : Ht && Ht in Object(t) ? En(t) : An(t);
}
function Re(t) {
  return t != null && typeof t == "object";
}
var Mn = "[object Symbol]";
function Ct(t) {
  return typeof t == "symbol" || Re(t) && ze(t) == Mn;
}
function Cn(t, e) {
  for (var i = -1, n = t == null ? 0 : t.length, r = Array(n); ++i < n; )
    r[i] = e(t[i], i, t);
  return r;
}
var fe = Array.isArray, Ut = ye ? ye.prototype : void 0, Vt = Ut ? Ut.toString : void 0;
function Wi(t) {
  if (typeof t == "string")
    return t;
  if (fe(t))
    return Cn(t, Wi) + "";
  if (Ct(t))
    return Vt ? Vt.call(t) : "";
  var e = t + "";
  return e == "0" && 1 / t == -1 / 0 ? "-0" : e;
}
function It(t) {
  var e = typeof t;
  return t != null && (e == "object" || e == "function");
}
function In(t) {
  return t;
}
var $n = "[object AsyncFunction]", Ln = "[object Function]", Rn = "[object GeneratorFunction]", Pn = "[object Proxy]";
function qi(t) {
  if (!It(t))
    return !1;
  var e = ze(t);
  return e == Ln || e == Rn || e == $n || e == Pn;
}
var mt = de["__core-js_shared__"], Wt = function() {
  var t = /[^.]+$/.exec(mt && mt.keys && mt.keys.IE_PROTO || "");
  return t ? "Symbol(src)_1." + t : "";
}();
function Nn(t) {
  return !!Wt && Wt in t;
}
var zn = Function.prototype, Fn = zn.toString;
function Te(t) {
  if (t != null) {
    try {
      return Fn.call(t);
    } catch {
    }
    try {
      return t + "";
    } catch {
    }
  }
  return "";
}
var Yn = /[\\^$.*+?()[\]{}|]/g, jn = /^\[object .+?Constructor\]$/, kn = Function.prototype, Hn = Object.prototype, Un = kn.toString, Vn = Hn.hasOwnProperty, Wn = RegExp(
  "^" + Un.call(Vn).replace(Yn, "\\$&").replace(/hasOwnProperty|(function).*?(?=\\\()| for .+?(?=\\\])/g, "$1.*?") + "$"
);
function qn(t) {
  if (!It(t) || Nn(t))
    return !1;
  var e = qi(t) ? Wn : jn;
  return e.test(Te(t));
}
function Bn(t, e) {
  return t == null ? void 0 : t[e];
}
function Fe(t, e) {
  var i = Bn(t, e);
  return qn(i) ? i : void 0;
}
var bt = Fe(de, "WeakMap");
function Xn() {
}
function Kn(t, e, i, n) {
  for (var r = t.length, a = i + -1; ++a < r; )
    if (e(t[a], a, t))
      return a;
  return -1;
}
function Gn(t) {
  return t !== t;
}
function Zn(t, e, i) {
  for (var n = i - 1, r = t.length; ++n < r; )
    if (t[n] === e)
      return n;
  return -1;
}
function Jn(t, e, i) {
  return e === e ? Zn(t, e, i) : Kn(t, Gn, i);
}
function Qn(t, e) {
  var i = t == null ? 0 : t.length;
  return !!i && Jn(t, e, 0) > -1;
}
var er = 9007199254740991, tr = /^(?:0|[1-9]\d*)$/;
function Bi(t, e) {
  var i = typeof t;
  return e = e ?? er, !!e && (i == "number" || i != "symbol" && tr.test(t)) && t > -1 && t % 1 == 0 && t < e;
}
function Xi(t, e) {
  return t === e || t !== t && e !== e;
}
var ir = 9007199254740991;
function $t(t) {
  return typeof t == "number" && t > -1 && t % 1 == 0 && t <= ir;
}
function nr(t) {
  return t != null && $t(t.length) && !qi(t);
}
var rr = Object.prototype;
function ar(t) {
  var e = t && t.constructor, i = typeof e == "function" && e.prototype || rr;
  return t === i;
}
function sr(t, e) {
  for (var i = -1, n = Array(t); ++i < t; )
    n[i] = e(i);
  return n;
}
var or = "[object Arguments]";
function qt(t) {
  return Re(t) && ze(t) == or;
}
var Ki = Object.prototype, lr = Ki.hasOwnProperty, ur = Ki.propertyIsEnumerable, Gi = qt(/* @__PURE__ */ function() {
  return arguments;
}()) ? qt : function(t) {
  return Re(t) && lr.call(t, "callee") && !ur.call(t, "callee");
};
function cr() {
  return !1;
}
var Zi = typeof exports == "object" && exports && !exports.nodeType && exports, Bt = Zi && typeof module == "object" && module && !module.nodeType && module, fr = Bt && Bt.exports === Zi, Xt = fr ? de.Buffer : void 0, hr = Xt ? Xt.isBuffer : void 0, _t = hr || cr, dr = "[object Arguments]", vr = "[object Array]", gr = "[object Boolean]", mr = "[object Date]", pr = "[object Error]", wr = "[object Function]", yr = "[object Map]", br = "[object Number]", _r = "[object Object]", Sr = "[object RegExp]", Er = "[object Set]", Tr = "[object String]", Dr = "[object WeakMap]", Ar = "[object ArrayBuffer]", Or = "[object DataView]", xr = "[object Float32Array]", Mr = "[object Float64Array]", Cr = "[object Int8Array]", Ir = "[object Int16Array]", $r = "[object Int32Array]", Lr = "[object Uint8Array]", Rr = "[object Uint8ClampedArray]", Pr = "[object Uint16Array]", Nr = "[object Uint32Array]", V = {};
V[xr] = V[Mr] = V[Cr] = V[Ir] = V[$r] = V[Lr] = V[Rr] = V[Pr] = V[Nr] = !0;
V[dr] = V[vr] = V[Ar] = V[gr] = V[Or] = V[mr] = V[pr] = V[wr] = V[yr] = V[br] = V[_r] = V[Sr] = V[Er] = V[Tr] = V[Dr] = !1;
function zr(t) {
  return Re(t) && $t(t.length) && !!V[ze(t)];
}
function Fr(t) {
  return function(e) {
    return t(e);
  };
}
var Ji = typeof exports == "object" && exports && !exports.nodeType && exports, ke = Ji && typeof module == "object" && module && !module.nodeType && module, Yr = ke && ke.exports === Ji, pt = Yr && Ui.process, Kt = function() {
  try {
    var t = ke && ke.require && ke.require("util").types;
    return t || pt && pt.binding && pt.binding("util");
  } catch {
  }
}(), Gt = Kt && Kt.isTypedArray, Qi = Gt ? Fr(Gt) : zr, jr = Object.prototype, kr = jr.hasOwnProperty;
function Hr(t, e) {
  var i = fe(t), n = !i && Gi(t), r = !i && !n && _t(t), a = !i && !n && !r && Qi(t), s = i || n || r || a, u = s ? sr(t.length, String) : [], o = u.length;
  for (var f in t)
    kr.call(t, f) && !(s && // Safari 9 has enumerable `arguments.length` in strict mode.
    (f == "length" || // Node.js 0.10 has enumerable non-index properties on buffers.
    r && (f == "offset" || f == "parent") || // PhantomJS 2 has enumerable non-index properties on typed arrays.
    a && (f == "buffer" || f == "byteLength" || f == "byteOffset") || // Skip index properties.
    Bi(f, o))) && u.push(f);
  return u;
}
function Ur(t, e) {
  return function(i) {
    return t(e(i));
  };
}
var Vr = Ur(Object.keys, Object), Wr = Object.prototype, qr = Wr.hasOwnProperty;
function Br(t) {
  if (!ar(t))
    return Vr(t);
  var e = [];
  for (var i in Object(t))
    qr.call(t, i) && i != "constructor" && e.push(i);
  return e;
}
function en(t) {
  return nr(t) ? Hr(t) : Br(t);
}
var Xr = /\.|\[(?:[^[\]]*|(["'])(?:(?!\1)[^\\]|\\.)*?\1)\]/, Kr = /^\w*$/;
function Lt(t, e) {
  if (fe(t))
    return !1;
  var i = typeof t;
  return i == "number" || i == "symbol" || i == "boolean" || t == null || Ct(t) ? !0 : Kr.test(t) || !Xr.test(t) || e != null && t in Object(e);
}
var Ue = Fe(Object, "create");
function Gr() {
  this.__data__ = Ue ? Ue(null) : {}, this.size = 0;
}
function Zr(t) {
  var e = this.has(t) && delete this.__data__[t];
  return this.size -= e ? 1 : 0, e;
}
var Jr = "__lodash_hash_undefined__", Qr = Object.prototype, ea = Qr.hasOwnProperty;
function ta(t) {
  var e = this.__data__;
  if (Ue) {
    var i = e[t];
    return i === Jr ? void 0 : i;
  }
  return ea.call(e, t) ? e[t] : void 0;
}
var ia = Object.prototype, na = ia.hasOwnProperty;
function ra(t) {
  var e = this.__data__;
  return Ue ? e[t] !== void 0 : na.call(e, t);
}
var aa = "__lodash_hash_undefined__";
function sa(t, e) {
  var i = this.__data__;
  return this.size += this.has(t) ? 0 : 1, i[t] = Ue && e === void 0 ? aa : e, this;
}
function Ee(t) {
  var e = -1, i = t == null ? 0 : t.length;
  for (this.clear(); ++e < i; ) {
    var n = t[e];
    this.set(n[0], n[1]);
  }
}
Ee.prototype.clear = Gr;
Ee.prototype.delete = Zr;
Ee.prototype.get = ta;
Ee.prototype.has = ra;
Ee.prototype.set = sa;
function oa() {
  this.__data__ = [], this.size = 0;
}
function ot(t, e) {
  for (var i = t.length; i--; )
    if (Xi(t[i][0], e))
      return i;
  return -1;
}
var la = Array.prototype, ua = la.splice;
function ca(t) {
  var e = this.__data__, i = ot(e, t);
  if (i < 0)
    return !1;
  var n = e.length - 1;
  return i == n ? e.pop() : ua.call(e, i, 1), --this.size, !0;
}
function fa(t) {
  var e = this.__data__, i = ot(e, t);
  return i < 0 ? void 0 : e[i][1];
}
function ha(t) {
  return ot(this.__data__, t) > -1;
}
function da(t, e) {
  var i = this.__data__, n = ot(i, t);
  return n < 0 ? (++this.size, i.push([t, e])) : i[n][1] = e, this;
}
function ve(t) {
  var e = -1, i = t == null ? 0 : t.length;
  for (this.clear(); ++e < i; ) {
    var n = t[e];
    this.set(n[0], n[1]);
  }
}
ve.prototype.clear = oa;
ve.prototype.delete = ca;
ve.prototype.get = fa;
ve.prototype.has = ha;
ve.prototype.set = da;
var Ve = Fe(de, "Map");
function va() {
  this.size = 0, this.__data__ = {
    hash: new Ee(),
    map: new (Ve || ve)(),
    string: new Ee()
  };
}
function ga(t) {
  var e = typeof t;
  return e == "string" || e == "number" || e == "symbol" || e == "boolean" ? t !== "__proto__" : t === null;
}
function lt(t, e) {
  var i = t.__data__;
  return ga(e) ? i[typeof e == "string" ? "string" : "hash"] : i.map;
}
function ma(t) {
  var e = lt(this, t).delete(t);
  return this.size -= e ? 1 : 0, e;
}
function pa(t) {
  return lt(this, t).get(t);
}
function wa(t) {
  return lt(this, t).has(t);
}
function ya(t, e) {
  var i = lt(this, t), n = i.size;
  return i.set(t, e), this.size += i.size == n ? 0 : 1, this;
}
function ge(t) {
  var e = -1, i = t == null ? 0 : t.length;
  for (this.clear(); ++e < i; ) {
    var n = t[e];
    this.set(n[0], n[1]);
  }
}
ge.prototype.clear = va;
ge.prototype.delete = ma;
ge.prototype.get = pa;
ge.prototype.has = wa;
ge.prototype.set = ya;
var ba = "Expected a function";
function Rt(t, e) {
  if (typeof t != "function" || e != null && typeof e != "function")
    throw new TypeError(ba);
  var i = function() {
    var n = arguments, r = e ? e.apply(this, n) : n[0], a = i.cache;
    if (a.has(r))
      return a.get(r);
    var s = t.apply(this, n);
    return i.cache = a.set(r, s) || a, s;
  };
  return i.cache = new (Rt.Cache || ge)(), i;
}
Rt.Cache = ge;
var _a = 500;
function Sa(t) {
  var e = Rt(t, function(n) {
    return i.size === _a && i.clear(), n;
  }), i = e.cache;
  return e;
}
var Ea = /[^.[\]]+|\[(?:(-?\d+(?:\.\d+)?)|(["'])((?:(?!\2)[^\\]|\\.)*?)\2)\]|(?=(?:\.|\[\])(?:\.|\[\]|$))/g, Ta = /\\(\\)?/g, Da = Sa(function(t) {
  var e = [];
  return t.charCodeAt(0) === 46 && e.push(""), t.replace(Ea, function(i, n, r, a) {
    e.push(r ? a.replace(Ta, "$1") : n || i);
  }), e;
});
function Aa(t) {
  return t == null ? "" : Wi(t);
}
function tn(t, e) {
  return fe(t) ? t : Lt(t, e) ? [t] : Da(Aa(t));
}
function ut(t) {
  if (typeof t == "string" || Ct(t))
    return t;
  var e = t + "";
  return e == "0" && 1 / t == -1 / 0 ? "-0" : e;
}
function nn(t, e) {
  e = tn(e, t);
  for (var i = 0, n = e.length; t != null && i < n; )
    t = t[ut(e[i++])];
  return i && i == n ? t : void 0;
}
function Oa(t, e, i) {
  var n = t == null ? void 0 : nn(t, e);
  return n === void 0 ? i : n;
}
function xa(t, e) {
  for (var i = -1, n = e.length, r = t.length; ++i < n; )
    t[r + i] = e[i];
  return t;
}
function Ma() {
  this.__data__ = new ve(), this.size = 0;
}
function Ca(t) {
  var e = this.__data__, i = e.delete(t);
  return this.size = e.size, i;
}
function Ia(t) {
  return this.__data__.get(t);
}
function $a(t) {
  return this.__data__.has(t);
}
var La = 200;
function Ra(t, e) {
  var i = this.__data__;
  if (i instanceof ve) {
    var n = i.__data__;
    if (!Ve || n.length < La - 1)
      return n.push([t, e]), this.size = ++i.size, this;
    i = this.__data__ = new ge(n);
  }
  return i.set(t, e), this.size = i.size, this;
}
function ce(t) {
  var e = this.__data__ = new ve(t);
  this.size = e.size;
}
ce.prototype.clear = Ma;
ce.prototype.delete = Ca;
ce.prototype.get = Ia;
ce.prototype.has = $a;
ce.prototype.set = Ra;
function Pa(t, e) {
  for (var i = -1, n = t == null ? 0 : t.length, r = 0, a = []; ++i < n; ) {
    var s = t[i];
    e(s, i, t) && (a[r++] = s);
  }
  return a;
}
function Na() {
  return [];
}
var za = Object.prototype, Fa = za.propertyIsEnumerable, Zt = Object.getOwnPropertySymbols, Ya = Zt ? function(t) {
  return t == null ? [] : (t = Object(t), Pa(Zt(t), function(e) {
    return Fa.call(t, e);
  }));
} : Na;
function ja(t, e, i) {
  var n = e(t);
  return fe(t) ? n : xa(n, i(t));
}
function Jt(t) {
  return ja(t, en, Ya);
}
var St = Fe(de, "DataView"), Et = Fe(de, "Promise"), Ie = Fe(de, "Set"), Qt = "[object Map]", ka = "[object Object]", ei = "[object Promise]", ti = "[object Set]", ii = "[object WeakMap]", ni = "[object DataView]", Ha = Te(St), Ua = Te(Ve), Va = Te(Et), Wa = Te(Ie), qa = Te(bt), we = ze;
(St && we(new St(new ArrayBuffer(1))) != ni || Ve && we(new Ve()) != Qt || Et && we(Et.resolve()) != ei || Ie && we(new Ie()) != ti || bt && we(new bt()) != ii) && (we = function(t) {
  var e = ze(t), i = e == ka ? t.constructor : void 0, n = i ? Te(i) : "";
  if (n)
    switch (n) {
      case Ha:
        return ni;
      case Ua:
        return Qt;
      case Va:
        return ei;
      case Wa:
        return ti;
      case qa:
        return ii;
    }
  return e;
});
var ri = de.Uint8Array, Ba = "__lodash_hash_undefined__";
function Xa(t) {
  return this.__data__.set(t, Ba), this;
}
function Ka(t) {
  return this.__data__.has(t);
}
function We(t) {
  var e = -1, i = t == null ? 0 : t.length;
  for (this.__data__ = new ge(); ++e < i; )
    this.add(t[e]);
}
We.prototype.add = We.prototype.push = Xa;
We.prototype.has = Ka;
function Ga(t, e) {
  for (var i = -1, n = t == null ? 0 : t.length; ++i < n; )
    if (e(t[i], i, t))
      return !0;
  return !1;
}
function rn(t, e) {
  return t.has(e);
}
var Za = 1, Ja = 2;
function an(t, e, i, n, r, a) {
  var s = i & Za, u = t.length, o = e.length;
  if (u != o && !(s && o > u))
    return !1;
  var f = a.get(t), c = a.get(e);
  if (f && c)
    return f == e && c == t;
  var l = -1, d = !0, b = i & Ja ? new We() : void 0;
  for (a.set(t, e), a.set(e, t); ++l < u; ) {
    var h = t[l], E = e[l];
    if (n)
      var g = s ? n(E, h, l, e, t, a) : n(h, E, l, t, e, a);
    if (g !== void 0) {
      if (g)
        continue;
      d = !1;
      break;
    }
    if (b) {
      if (!Ga(e, function(m, _) {
        if (!rn(b, _) && (h === m || r(h, m, i, n, a)))
          return b.push(_);
      })) {
        d = !1;
        break;
      }
    } else if (!(h === E || r(h, E, i, n, a))) {
      d = !1;
      break;
    }
  }
  return a.delete(t), a.delete(e), d;
}
function Qa(t) {
  var e = -1, i = Array(t.size);
  return t.forEach(function(n, r) {
    i[++e] = [r, n];
  }), i;
}
function Pt(t) {
  var e = -1, i = Array(t.size);
  return t.forEach(function(n) {
    i[++e] = n;
  }), i;
}
var es = 1, ts = 2, is = "[object Boolean]", ns = "[object Date]", rs = "[object Error]", as = "[object Map]", ss = "[object Number]", os = "[object RegExp]", ls = "[object Set]", us = "[object String]", cs = "[object Symbol]", fs = "[object ArrayBuffer]", hs = "[object DataView]", ai = ye ? ye.prototype : void 0, wt = ai ? ai.valueOf : void 0;
function ds(t, e, i, n, r, a, s) {
  switch (i) {
    case hs:
      if (t.byteLength != e.byteLength || t.byteOffset != e.byteOffset)
        return !1;
      t = t.buffer, e = e.buffer;
    case fs:
      return !(t.byteLength != e.byteLength || !a(new ri(t), new ri(e)));
    case is:
    case ns:
    case ss:
      return Xi(+t, +e);
    case rs:
      return t.name == e.name && t.message == e.message;
    case os:
    case us:
      return t == e + "";
    case as:
      var u = Qa;
    case ls:
      var o = n & es;
      if (u || (u = Pt), t.size != e.size && !o)
        return !1;
      var f = s.get(t);
      if (f)
        return f == e;
      n |= ts, s.set(t, e);
      var c = an(u(t), u(e), n, r, a, s);
      return s.delete(t), c;
    case cs:
      if (wt)
        return wt.call(t) == wt.call(e);
  }
  return !1;
}
var vs = 1, gs = Object.prototype, ms = gs.hasOwnProperty;
function ps(t, e, i, n, r, a) {
  var s = i & vs, u = Jt(t), o = u.length, f = Jt(e), c = f.length;
  if (o != c && !s)
    return !1;
  for (var l = o; l--; ) {
    var d = u[l];
    if (!(s ? d in e : ms.call(e, d)))
      return !1;
  }
  var b = a.get(t), h = a.get(e);
  if (b && h)
    return b == e && h == t;
  var E = !0;
  a.set(t, e), a.set(e, t);
  for (var g = s; ++l < o; ) {
    d = u[l];
    var m = t[d], _ = e[d];
    if (n)
      var S = s ? n(_, m, d, e, t, a) : n(m, _, d, t, e, a);
    if (!(S === void 0 ? m === _ || r(m, _, i, n, a) : S)) {
      E = !1;
      break;
    }
    g || (g = d == "constructor");
  }
  if (E && !g) {
    var P = t.constructor, N = e.constructor;
    P != N && "constructor" in t && "constructor" in e && !(typeof P == "function" && P instanceof P && typeof N == "function" && N instanceof N) && (E = !1);
  }
  return a.delete(t), a.delete(e), E;
}
var ws = 1, si = "[object Arguments]", oi = "[object Array]", Ze = "[object Object]", ys = Object.prototype, li = ys.hasOwnProperty;
function bs(t, e, i, n, r, a) {
  var s = fe(t), u = fe(e), o = s ? oi : we(t), f = u ? oi : we(e);
  o = o == si ? Ze : o, f = f == si ? Ze : f;
  var c = o == Ze, l = f == Ze, d = o == f;
  if (d && _t(t)) {
    if (!_t(e))
      return !1;
    s = !0, c = !1;
  }
  if (d && !c)
    return a || (a = new ce()), s || Qi(t) ? an(t, e, i, n, r, a) : ds(t, e, o, i, n, r, a);
  if (!(i & ws)) {
    var b = c && li.call(t, "__wrapped__"), h = l && li.call(e, "__wrapped__");
    if (b || h) {
      var E = b ? t.value() : t, g = h ? e.value() : e;
      return a || (a = new ce()), r(E, g, i, n, a);
    }
  }
  return d ? (a || (a = new ce()), ps(t, e, i, n, r, a)) : !1;
}
function Nt(t, e, i, n, r) {
  return t === e ? !0 : t == null || e == null || !Re(t) && !Re(e) ? t !== t && e !== e : bs(t, e, i, n, Nt, r);
}
var _s = 1, Ss = 2;
function Es(t, e, i, n) {
  var r = i.length, a = r;
  if (t == null)
    return !a;
  for (t = Object(t); r--; ) {
    var s = i[r];
    if (s[2] ? s[1] !== t[s[0]] : !(s[0] in t))
      return !1;
  }
  for (; ++r < a; ) {
    s = i[r];
    var u = s[0], o = t[u], f = s[1];
    if (s[2]) {
      if (o === void 0 && !(u in t))
        return !1;
    } else {
      var c = new ce(), l;
      if (!(l === void 0 ? Nt(f, o, _s | Ss, n, c) : l))
        return !1;
    }
  }
  return !0;
}
function sn(t) {
  return t === t && !It(t);
}
function Ts(t) {
  for (var e = en(t), i = e.length; i--; ) {
    var n = e[i], r = t[n];
    e[i] = [n, r, sn(r)];
  }
  return e;
}
function on(t, e) {
  return function(i) {
    return i == null ? !1 : i[t] === e && (e !== void 0 || t in Object(i));
  };
}
function Ds(t) {
  var e = Ts(t);
  return e.length == 1 && e[0][2] ? on(e[0][0], e[0][1]) : function(i) {
    return i === t || Es(i, t, e);
  };
}
function As(t, e) {
  return t != null && e in Object(t);
}
function Os(t, e, i) {
  e = tn(e, t);
  for (var n = -1, r = e.length, a = !1; ++n < r; ) {
    var s = ut(e[n]);
    if (!(a = t != null && i(t, s)))
      break;
    t = t[s];
  }
  return a || ++n != r ? a : (r = t == null ? 0 : t.length, !!r && $t(r) && Bi(s, r) && (fe(t) || Gi(t)));
}
function xs(t, e) {
  return t != null && Os(t, e, As);
}
var Ms = 1, Cs = 2;
function Is(t, e) {
  return Lt(t) && sn(e) ? on(ut(t), e) : function(i) {
    var n = Oa(i, t);
    return n === void 0 && n === e ? xs(i, t) : Nt(e, n, Ms | Cs);
  };
}
function $s(t) {
  return function(e) {
    return e == null ? void 0 : e[t];
  };
}
function Ls(t) {
  return function(e) {
    return nn(e, t);
  };
}
function Rs(t) {
  return Lt(t) ? $s(ut(t)) : Ls(t);
}
function Ps(t) {
  return typeof t == "function" ? t : t == null ? In : typeof t == "object" ? fe(t) ? Is(t[0], t[1]) : Ds(t) : Rs(t);
}
var Ns = 1 / 0, zs = Ie && 1 / Pt(new Ie([, -0]))[1] == Ns ? function(t) {
  return new Ie(t);
} : Xn, Fs = 200;
function Ys(t, e, i) {
  var n = -1, r = Qn, a = t.length, s = !0, u = [], o = u;
  if (a >= Fs) {
    var f = e ? null : zs(t);
    if (f)
      return Pt(f);
    s = !1, r = rn, o = new We();
  } else
    o = e ? [] : u;
  e:
    for (; ++n < a; ) {
      var c = t[n], l = e ? e(c) : c;
      if (c = c !== 0 ? c : 0, s && l === l) {
        for (var d = o.length; d--; )
          if (o[d] === l)
            continue e;
        e && o.push(l), u.push(c);
      } else r(o, l, i) || (o !== u && o.push(l), u.push(c));
    }
  return u;
}
function js(t, e) {
  return t && t.length ? Ys(t, Ps(e)) : [];
}
mn.extend(wn);
function ks() {
  return Number(Math.random().toString().substr(3, 4) + Date.now()).toString(36);
}
const Co = (t) => t == null || t === "";
function Io(t) {
  if (t === 0) return "0 B";
  const e = 1024, i = ["B", "KB", "MB", "GB", "TB"], n = Math.floor(Math.log(t) / Math.log(e));
  return parseFloat((t / Math.pow(e, n)).toFixed(2)) + " " + i[n];
}
const Hs = (t, e = "", i) => {
  const n = new Blob([t]);
  if (window.navigator.msSaveOrOpenBlob)
    navigator.msSaveBlob(n, e);
  else {
    const r = document.createElement("a");
    i && (r.className = i), r.download = e, r.style.display = "none", r.href = URL.createObjectURL(n), document.body.appendChild(r), r.click(), document.body.removeChild(r);
  }
}, Us = (t, e, i) => {
  const { data: n, headers: r } = t;
  let a = "";
  const s = r["content-disposition"];
  s && (a = s == null ? void 0 : s.substring(s.indexOf("filename=") + 9, s.length), a = decodeURI(a), a = a.replace(/"/g, ""));
  const u = e || a, o = new FileReader();
  o.onload = (f) => {
    try {
      const c = JSON.parse((f == null ? void 0 : f.target).result);
      c.success || it.error(c.errMessage);
    } catch {
      Hs(n, u, i);
    }
  }, o.readAsText(n);
}, $o = (t) => {
  const e = document.createElement("form");
  e.setAttribute("method", "get"), e.setAttribute("action", t), document.body.appendChild(e), e.submit(), document.body.removeChild(e);
}, Lo = (t, e, i = "") => {
  const n = new Blob([t], {
    type: e
  }), r = new FileReader();
  r.readAsDataURL(n), r.onload = function(a) {
    var u;
    const s = document.createElement("a");
    s.download = i || `${(/* @__PURE__ */ new Date()).getTime()}`, s.href = (u = a.target) == null ? void 0 : u.result, document.body.appendChild(s), s.click(), document.body.removeChild(s);
  };
};
/*!
 * Viewer.js v1.11.7
 * https://fengyuanchen.github.io/viewerjs
 *
 * Copyright 2015-present Chen Fengyuan
 * Released under the MIT license
 *
 * Date: 2024-11-24T04:32:19.116Z
 */
function Vs(t, e) {
  if (!(t instanceof e)) throw new TypeError("Cannot call a class as a function");
}
function ui(t, e) {
  for (var i = 0; i < e.length; i++) {
    var n = e[i];
    n.enumerable = n.enumerable || !1, n.configurable = !0, "value" in n && (n.writable = !0), Object.defineProperty(t, ln(n.key), n);
  }
}
function Ws(t, e, i) {
  return e && ui(t.prototype, e), i && ui(t, i), Object.defineProperty(t, "prototype", {
    writable: !1
  }), t;
}
function qs(t, e, i) {
  return (e = ln(e)) in t ? Object.defineProperty(t, e, {
    value: i,
    enumerable: !0,
    configurable: !0,
    writable: !0
  }) : t[e] = i, t;
}
function ci(t, e) {
  var i = Object.keys(t);
  if (Object.getOwnPropertySymbols) {
    var n = Object.getOwnPropertySymbols(t);
    e && (n = n.filter(function(r) {
      return Object.getOwnPropertyDescriptor(t, r).enumerable;
    })), i.push.apply(i, n);
  }
  return i;
}
function zt(t) {
  for (var e = 1; e < arguments.length; e++) {
    var i = arguments[e] != null ? arguments[e] : {};
    e % 2 ? ci(Object(i), !0).forEach(function(n) {
      qs(t, n, i[n]);
    }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(t, Object.getOwnPropertyDescriptors(i)) : ci(Object(i)).forEach(function(n) {
      Object.defineProperty(t, n, Object.getOwnPropertyDescriptor(i, n));
    });
  }
  return t;
}
function Bs(t, e) {
  if (typeof t != "object" || !t) return t;
  var i = t[Symbol.toPrimitive];
  if (i !== void 0) {
    var n = i.call(t, e);
    if (typeof n != "object") return n;
    throw new TypeError("@@toPrimitive must return a primitive value.");
  }
  return String(t);
}
function ln(t) {
  var e = Bs(t, "string");
  return typeof e == "symbol" ? e : e + "";
}
function Tt(t) {
  "@babel/helpers - typeof";
  return Tt = typeof Symbol == "function" && typeof Symbol.iterator == "symbol" ? function(e) {
    return typeof e;
  } : function(e) {
    return e && typeof Symbol == "function" && e.constructor === Symbol && e !== Symbol.prototype ? "symbol" : typeof e;
  }, Tt(t);
}
var fi = {
  /**
   * Enable a modal backdrop, specify `static` for a backdrop
   * which doesn't close the modal on click.
   * @type {boolean}
   */
  backdrop: !0,
  /**
   * Show the button on the top-right of the viewer.
   * @type {boolean}
   */
  button: !0,
  /**
   * Show the navbar.
   * @type {boolean | number}
   */
  navbar: !0,
  /**
   * Specify the visibility and the content of the title.
   * @type {boolean | number | Function | Array}
   */
  title: !0,
  /**
   * Show the toolbar.
   * @type {boolean | number | Object}
   */
  toolbar: !0,
  /**
   * Custom class name(s) to add to the viewer's root element.
   * @type {string}
   */
  className: "",
  /**
   * Define where to put the viewer in modal mode.
   * @type {string | Element}
   */
  container: "body",
  /**
   * Filter the images for viewing. Return true if the image is viewable.
   * @type {Function}
   */
  filter: null,
  /**
   * Enable to request fullscreen when play.
   * {@link https://developer.mozilla.org/en-US/docs/Web/API/FullscreenOptions}
   * @type {boolean|FullscreenOptions}
   */
  fullscreen: !0,
  /**
   * Define the extra attributes to inherit from the original image.
   * @type {Array}
   */
  inheritedAttributes: ["crossOrigin", "decoding", "isMap", "loading", "referrerPolicy", "sizes", "srcset", "useMap"],
  /**
   * Define the initial coverage of the viewing image.
   * @type {number}
   */
  initialCoverage: 0.9,
  /**
   * Define the initial index of the image for viewing.
   * @type {number}
   */
  initialViewIndex: 0,
  /**
   * Enable inline mode.
   * @type {boolean}
   */
  inline: !1,
  /**
   * The amount of time to delay between automatically cycling an image when playing.
   * @type {number}
   */
  interval: 5e3,
  /**
   * Enable keyboard support.
   * @type {boolean}
   */
  keyboard: !0,
  /**
   * Focus the viewer when initialized.
   * @type {boolean}
   */
  focus: !0,
  /**
   * Indicate if show a loading spinner when load image or not.
   * @type {boolean}
   */
  loading: !0,
  /**
   * Indicate if enable loop viewing or not.
   * @type {boolean}
   */
  loop: !0,
  /**
   * Min width of the viewer in inline mode.
   * @type {number}
   */
  minWidth: 200,
  /**
   * Min height of the viewer in inline mode.
   * @type {number}
   */
  minHeight: 100,
  /**
   * Enable to move the image.
   * @type {boolean}
   */
  movable: !0,
  /**
   * Enable to rotate the image.
   * @type {boolean}
   */
  rotatable: !0,
  /**
   * Enable to scale the image.
   * @type {boolean}
   */
  scalable: !0,
  /**
   * Enable to zoom the image.
   * @type {boolean}
   */
  zoomable: !0,
  /**
   * Enable to zoom the current image by dragging on the touch screen.
   * @type {boolean}
   */
  zoomOnTouch: !0,
  /**
   * Enable to zoom the image by wheeling mouse.
   * @type {boolean}
   */
  zoomOnWheel: !0,
  /**
   * Enable to slide to the next or previous image by swiping on the touch screen.
   * @type {boolean}
   */
  slideOnTouch: !0,
  /**
   * Indicate if toggle the image size between its natural size
   * and initial size when double click on the image or not.
   * @type {boolean}
   */
  toggleOnDblclick: !0,
  /**
   * Show the tooltip with image ratio (percentage) when zoom in or zoom out.
   * @type {boolean}
   */
  tooltip: !0,
  /**
   * Enable CSS3 Transition for some special elements.
   * @type {boolean}
   */
  transition: !0,
  /**
   * Define the CSS `z-index` value of viewer in modal mode.
   * @type {number}
   */
  zIndex: 2015,
  /**
   * Define the CSS `z-index` value of viewer in inline mode.
   * @type {number}
   */
  zIndexInline: 0,
  /**
   * Define the ratio when zoom the image by wheeling mouse.
   * @type {number}
   */
  zoomRatio: 0.1,
  /**
   * Define the min ratio of the image when zoom out.
   * @type {number}
   */
  minZoomRatio: 0.01,
  /**
   * Define the max ratio of the image when zoom in.
   * @type {number}
   */
  maxZoomRatio: 100,
  /**
   * Define where to get the original image URL for viewing.
   * @type {string | Function}
   */
  url: "src",
  /**
   * Event shortcuts.
   * @type {Function}
   */
  ready: null,
  show: null,
  shown: null,
  hide: null,
  hidden: null,
  view: null,
  viewed: null,
  move: null,
  moved: null,
  rotate: null,
  rotated: null,
  scale: null,
  scaled: null,
  zoom: null,
  zoomed: null,
  play: null,
  stop: null
}, Xs = '<div class="viewer-container" tabindex="-1" touch-action="none"><div class="viewer-canvas"></div><div class="viewer-footer"><div class="viewer-title"></div><div class="viewer-toolbar"></div><div class="viewer-navbar"><ul class="viewer-list" role="navigation"></ul></div></div><div class="viewer-tooltip" role="alert" aria-hidden="true"></div><div class="viewer-button" data-viewer-action="mix" role="button"></div><div class="viewer-player"></div></div>', ct = typeof window < "u" && typeof window.document < "u", he = ct ? window : {}, $e = ct && he.document.documentElement ? "ontouchstart" in he.document.documentElement : !1, Ft = ct ? "PointerEvent" in he : !1, L = "viewer", nt = "move", un = "switch", je = "zoom", Je = "".concat(L, "-active"), Ks = "".concat(L, "-close"), rt = "".concat(L, "-fade"), Dt = "".concat(L, "-fixed"), Gs = "".concat(L, "-fullscreen"), hi = "".concat(L, "-fullscreen-exit"), _e = "".concat(L, "-hide"), Zs = "".concat(L, "-hide-md-down"), Js = "".concat(L, "-hide-sm-down"), Qs = "".concat(L, "-hide-xs-down"), ne = "".concat(L, "-in"), He = "".concat(L, "-invisible"), Le = "".concat(L, "-loading"), eo = "".concat(L, "-move"), di = "".concat(L, "-open"), De = "".concat(L, "-show"), K = "".concat(L, "-transition"), Pe = "click", At = "dblclick", vi = "dragstart", gi = "focusin", mi = "keydown", re = "load", Se = "error", to = $e ? "touchend touchcancel" : "mouseup", io = $e ? "touchmove" : "mousemove", no = $e ? "touchstart" : "mousedown", pi = Ft ? "pointerdown" : no, wi = Ft ? "pointermove" : io, yi = Ft ? "pointerup pointercancel" : to, bi = "resize", ae = "transitionend", _i = "wheel", Si = "ready", Ei = "show", Ti = "shown", Di = "hide", Ai = "hidden", Oi = "view", qe = "viewed", xi = "move", Mi = "moved", Ci = "rotate", Ii = "rotated", $i = "scale", Li = "scaled", Ri = "zoom", Pi = "zoomed", Ni = "play", zi = "stop", st = "".concat(L, "Action"), Yt = /\s\s*/, Qe = ["zoom-in", "zoom-out", "one-to-one", "reset", "prev", "play", "next", "rotate-left", "rotate-right", "flip-horizontal", "flip-vertical"];
function Be(t) {
  return typeof t == "string";
}
var ro = Number.isNaN || he.isNaN;
function B(t) {
  return typeof t == "number" && !ro(t);
}
function xe(t) {
  return typeof t > "u";
}
function Ne(t) {
  return Tt(t) === "object" && t !== null;
}
var ao = Object.prototype.hasOwnProperty;
function Me(t) {
  if (!Ne(t))
    return !1;
  try {
    var e = t.constructor, i = e.prototype;
    return e && i && ao.call(i, "isPrototypeOf");
  } catch {
    return !1;
  }
}
function Y(t) {
  return typeof t == "function";
}
function H(t, e) {
  if (t && Y(e))
    if (Array.isArray(t) || B(t.length)) {
      var i = t.length, n;
      for (n = 0; n < i && e.call(t, t[n], n, t) !== !1; n += 1)
        ;
    } else Ne(t) && Object.keys(t).forEach(function(r) {
      e.call(t, t[r], r, t);
    });
  return t;
}
var te = Object.assign || function(e) {
  for (var i = arguments.length, n = new Array(i > 1 ? i - 1 : 0), r = 1; r < i; r++)
    n[r - 1] = arguments[r];
  return Ne(e) && n.length > 0 && n.forEach(function(a) {
    Ne(a) && Object.keys(a).forEach(function(s) {
      e[s] = a[s];
    });
  }), e;
}, so = /^(?:width|height|left|top|marginLeft|marginTop)$/;
function se(t, e) {
  var i = t.style;
  H(e, function(n, r) {
    so.test(r) && B(n) && (n += "px"), i[r] = n;
  });
}
function oo(t) {
  return Be(t) ? t.replace(/&(?!amp;|quot;|#39;|lt;|gt;)/g, "&amp;").replace(/"/g, "&quot;").replace(/'/g, "&#39;").replace(/</g, "&lt;").replace(/>/g, "&gt;") : t;
}
function Ae(t, e) {
  return !t || !e ? !1 : t.classList ? t.classList.contains(e) : t.className.indexOf(e) > -1;
}
function M(t, e) {
  if (!(!t || !e)) {
    if (B(t.length)) {
      H(t, function(n) {
        M(n, e);
      });
      return;
    }
    if (t.classList) {
      t.classList.add(e);
      return;
    }
    var i = t.className.trim();
    i ? i.indexOf(e) < 0 && (t.className = "".concat(i, " ").concat(e)) : t.className = e;
  }
}
function z(t, e) {
  if (!(!t || !e)) {
    if (B(t.length)) {
      H(t, function(i) {
        z(i, e);
      });
      return;
    }
    if (t.classList) {
      t.classList.remove(e);
      return;
    }
    t.className.indexOf(e) >= 0 && (t.className = t.className.replace(e, ""));
  }
}
function Xe(t, e, i) {
  if (e) {
    if (B(t.length)) {
      H(t, function(n) {
        Xe(n, e, i);
      });
      return;
    }
    i ? M(t, e) : z(t, e);
  }
}
var lo = /([a-z\d])([A-Z])/g;
function jt(t) {
  return t.replace(lo, "$1-$2").toLowerCase();
}
function Ce(t, e) {
  return Ne(t[e]) ? t[e] : t.dataset ? t.dataset[e] : t.getAttribute("data-".concat(jt(e)));
}
function Ot(t, e, i) {
  Ne(i) ? t[e] = i : t.dataset ? t.dataset[e] = i : t.setAttribute("data-".concat(jt(e)), i);
}
var cn = function() {
  var t = !1;
  if (ct) {
    var e = !1, i = function() {
    }, n = Object.defineProperty({}, "once", {
      get: function() {
        return t = !0, e;
      },
      /**
       * This setter can fix a `TypeError` in strict mode
       * {@link https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Errors/Getter_only}
       * @param {boolean} value - The value to set
       */
      set: function(a) {
        e = a;
      }
    });
    he.addEventListener("test", i, n), he.removeEventListener("test", i, n);
  }
  return t;
}();
function j(t, e, i) {
  var n = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : {}, r = i;
  e.trim().split(Yt).forEach(function(a) {
    if (!cn) {
      var s = t.listeners;
      s && s[a] && s[a][i] && (r = s[a][i], delete s[a][i], Object.keys(s[a]).length === 0 && delete s[a], Object.keys(s).length === 0 && delete t.listeners);
    }
    t.removeEventListener(a, r, n);
  });
}
function C(t, e, i) {
  var n = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : {}, r = i;
  e.trim().split(Yt).forEach(function(a) {
    if (n.once && !cn) {
      var s = t.listeners, u = s === void 0 ? {} : s;
      r = function() {
        delete u[a][i], t.removeEventListener(a, r, n);
        for (var f = arguments.length, c = new Array(f), l = 0; l < f; l++)
          c[l] = arguments[l];
        i.apply(t, c);
      }, u[a] || (u[a] = {}), u[a][i] && t.removeEventListener(a, u[a][i], n), u[a][i] = r, t.listeners = u;
    }
    t.addEventListener(a, r, n);
  });
}
function G(t, e, i, n) {
  var r;
  return Y(Event) && Y(CustomEvent) ? r = new CustomEvent(e, zt({
    bubbles: !0,
    cancelable: !0,
    detail: i
  }, n)) : (r = document.createEvent("CustomEvent"), r.initCustomEvent(e, !0, !0, i)), t.dispatchEvent(r);
}
function uo(t) {
  var e = t.getBoundingClientRect();
  return {
    left: e.left + (window.pageXOffset - document.documentElement.clientLeft),
    top: e.top + (window.pageYOffset - document.documentElement.clientTop)
  };
}
function at(t) {
  var e = t.rotate, i = t.scaleX, n = t.scaleY, r = t.translateX, a = t.translateY, s = [];
  B(r) && r !== 0 && s.push("translateX(".concat(r, "px)")), B(a) && a !== 0 && s.push("translateY(".concat(a, "px)")), B(e) && e !== 0 && s.push("rotate(".concat(e, "deg)")), B(i) && i !== 1 && s.push("scaleX(".concat(i, ")")), B(n) && n !== 1 && s.push("scaleY(".concat(n, ")"));
  var u = s.length ? s.join(" ") : "none";
  return {
    WebkitTransform: u,
    msTransform: u,
    transform: u
  };
}
function co(t) {
  return Be(t) ? decodeURIComponent(t.replace(/^.*\//, "").replace(/[?&#].*$/, "")) : "";
}
var yt = he.navigator && /Version\/\d+(\.\d+)+?\s+Safari/i.test(he.navigator.userAgent);
function fn(t, e, i) {
  var n = document.createElement("img");
  if (t.naturalWidth && !yt)
    return i(t.naturalWidth, t.naturalHeight), n;
  var r = document.body || document.documentElement;
  return n.onload = function() {
    i(n.width, n.height), yt || r.removeChild(n);
  }, H(e.inheritedAttributes, function(a) {
    var s = t.getAttribute(a);
    s !== null && n.setAttribute(a, s);
  }), n.src = t.src, yt || (n.style.cssText = "left:0;max-height:none!important;max-width:none!important;min-height:0!important;min-width:0!important;opacity:0;position:absolute;top:0;z-index:-1;", r.appendChild(n)), n;
}
function et(t) {
  switch (t) {
    case 2:
      return Qs;
    case 3:
      return Js;
    case 4:
      return Zs;
    default:
      return "";
  }
}
function fo(t) {
  var e = zt({}, t), i = [];
  return H(t, function(n, r) {
    delete e[r], H(e, function(a) {
      var s = Math.abs(n.startX - a.startX), u = Math.abs(n.startY - a.startY), o = Math.abs(n.endX - a.endX), f = Math.abs(n.endY - a.endY), c = Math.sqrt(s * s + u * u), l = Math.sqrt(o * o + f * f), d = (l - c) / c;
      i.push(d);
    });
  }), i.sort(function(n, r) {
    return Math.abs(n) < Math.abs(r);
  }), i[0];
}
function tt(t, e) {
  var i = t.pageX, n = t.pageY, r = {
    endX: i,
    endY: n
  };
  return e ? r : zt({
    timeStamp: Date.now(),
    startX: i,
    startY: n
  }, r);
}
function ho(t) {
  var e = 0, i = 0, n = 0;
  return H(t, function(r) {
    var a = r.startX, s = r.startY;
    e += a, i += s, n += 1;
  }), e /= n, i /= n, {
    pageX: e,
    pageY: i
  };
}
var vo = {
  render: function() {
    this.initContainer(), this.initViewer(), this.initList(), this.renderViewer();
  },
  initBody: function() {
    var e = this.element.ownerDocument, i = e.body || e.documentElement;
    this.body = i, this.scrollbarWidth = window.innerWidth - e.documentElement.clientWidth, this.initialBodyPaddingRight = i.style.paddingRight, this.initialBodyComputedPaddingRight = window.getComputedStyle(i).paddingRight;
  },
  initContainer: function() {
    this.containerData = {
      width: window.innerWidth,
      height: window.innerHeight
    };
  },
  initViewer: function() {
    var e = this.options, i = this.parent, n;
    e.inline && (n = {
      width: Math.max(i.offsetWidth, e.minWidth),
      height: Math.max(i.offsetHeight, e.minHeight)
    }, this.parentData = n), (this.fulled || !n) && (n = this.containerData), this.viewerData = te({}, n);
  },
  renderViewer: function() {
    this.options.inline && !this.fulled && se(this.viewer, this.viewerData);
  },
  initList: function() {
    var e = this, i = this.element, n = this.options, r = this.list, a = [];
    r.innerHTML = "", H(this.images, function(s, u) {
      var o = s.src, f = s.alt || co(o), c = e.getImageURL(s);
      if (o || c) {
        var l = document.createElement("li"), d = document.createElement("img");
        H(n.inheritedAttributes, function(b) {
          var h = s.getAttribute(b);
          h !== null && d.setAttribute(b, h);
        }), n.navbar && (d.src = o || c), d.alt = f, d.setAttribute("data-original-url", c || o), l.setAttribute("data-index", u), l.setAttribute("data-viewer-action", "view"), l.setAttribute("role", "button"), n.keyboard && l.setAttribute("tabindex", 0), l.appendChild(d), r.appendChild(l), a.push(l);
      }
    }), this.items = a, H(a, function(s) {
      var u = s.firstElementChild, o, f;
      Ot(u, "filled", !0), n.loading && M(s, Le), C(u, re, o = function(l) {
        j(u, Se, f), n.loading && z(s, Le), e.loadImage(l);
      }, {
        once: !0
      }), C(u, Se, f = function() {
        j(u, re, o), n.loading && z(s, Le);
      }, {
        once: !0
      });
    }), n.transition && C(i, qe, function() {
      M(r, K);
    }, {
      once: !0
    });
  },
  renderList: function() {
    var e = this.index, i = this.items[e];
    if (i) {
      var n = i.nextElementSibling, r = parseInt(window.getComputedStyle(n || i).marginLeft, 10), a = i.offsetWidth, s = a + r;
      se(this.list, te({
        width: s * this.length - r
      }, at({
        translateX: (this.viewerData.width - a) / 2 - s * e
      })));
    }
  },
  resetList: function() {
    var e = this.list;
    e.innerHTML = "", z(e, K), se(e, at({
      translateX: 0
    }));
  },
  initImage: function(e) {
    var i = this, n = this.options, r = this.image, a = this.viewerData, s = this.footer.offsetHeight, u = a.width, o = Math.max(a.height - s, s), f = this.imageData || {}, c;
    this.imageInitializing = {
      abort: function() {
        c.onload = null;
      }
    }, c = fn(r, n, function(l, d) {
      var b = l / d, h = Math.max(0, Math.min(1, n.initialCoverage)), E = u, g = o;
      i.imageInitializing = !1, o * b > u ? g = u / b : E = o * b, h = B(h) ? h : 0.9, E = Math.min(E * h, l), g = Math.min(g * h, d);
      var m = (u - E) / 2, _ = (o - g) / 2, S = {
        left: m,
        top: _,
        x: m,
        y: _,
        width: E,
        height: g,
        oldRatio: 1,
        ratio: E / l,
        aspectRatio: b,
        naturalWidth: l,
        naturalHeight: d
      }, P = te({}, S);
      n.rotatable && (S.rotate = f.rotate || 0, P.rotate = 0), n.scalable && (S.scaleX = f.scaleX || 1, S.scaleY = f.scaleY || 1, P.scaleX = 1, P.scaleY = 1), i.imageData = S, i.initialImageData = P, e && e();
    });
  },
  renderImage: function(e) {
    var i = this, n = this.image, r = this.imageData;
    if (se(n, te({
      width: r.width,
      height: r.height,
      // XXX: Not to use translateX/Y to avoid image shaking when zooming
      marginLeft: r.x,
      marginTop: r.y
    }, at(r))), e)
      if ((this.viewing || this.moving || this.rotating || this.scaling || this.zooming) && this.options.transition && Ae(n, K)) {
        var a = function() {
          i.imageRendering = !1, e();
        };
        this.imageRendering = {
          abort: function() {
            j(n, ae, a);
          }
        }, C(n, ae, a, {
          once: !0
        });
      } else
        e();
  },
  resetImage: function() {
    var e = this.image;
    e && (this.viewing && this.viewing.abort(), e.parentNode.removeChild(e), this.image = null, this.title.innerHTML = "");
  }
}, go = {
  bind: function() {
    var e = this.options, i = this.viewer, n = this.canvas, r = this.element.ownerDocument;
    C(i, Pe, this.onClick = this.click.bind(this)), C(i, vi, this.onDragStart = this.dragstart.bind(this)), C(n, pi, this.onPointerDown = this.pointerdown.bind(this)), C(r, wi, this.onPointerMove = this.pointermove.bind(this)), C(r, yi, this.onPointerUp = this.pointerup.bind(this)), C(r, mi, this.onKeyDown = this.keydown.bind(this)), C(window, bi, this.onResize = this.resize.bind(this)), e.zoomable && e.zoomOnWheel && C(i, _i, this.onWheel = this.wheel.bind(this), {
      passive: !1,
      capture: !0
    }), e.toggleOnDblclick && C(n, At, this.onDblclick = this.dblclick.bind(this));
  },
  unbind: function() {
    var e = this.options, i = this.viewer, n = this.canvas, r = this.element.ownerDocument;
    j(i, Pe, this.onClick), j(i, vi, this.onDragStart), j(n, pi, this.onPointerDown), j(r, wi, this.onPointerMove), j(r, yi, this.onPointerUp), j(r, mi, this.onKeyDown), j(window, bi, this.onResize), e.zoomable && e.zoomOnWheel && j(i, _i, this.onWheel, {
      passive: !1,
      capture: !0
    }), e.toggleOnDblclick && j(n, At, this.onDblclick);
  }
}, mo = {
  click: function(e) {
    var i = this.options, n = this.imageData, r = e.target, a = Ce(r, st);
    switch (!a && r.localName === "img" && r.parentElement.localName === "li" && (r = r.parentElement, a = Ce(r, st)), $e && e.isTrusted && r === this.canvas && clearTimeout(this.clickCanvasTimeout), a) {
      case "mix":
        this.played ? this.stop() : i.inline ? this.fulled ? this.exit() : this.full() : this.hide();
        break;
      case "hide":
        this.pointerMoved || this.hide();
        break;
      case "view":
        this.view(Ce(r, "index"));
        break;
      case "zoom-in":
        this.zoom(0.1, !0);
        break;
      case "zoom-out":
        this.zoom(-0.1, !0);
        break;
      case "one-to-one":
        this.toggle();
        break;
      case "reset":
        this.reset();
        break;
      case "prev":
        this.prev(i.loop);
        break;
      case "play":
        this.play(i.fullscreen);
        break;
      case "next":
        this.next(i.loop);
        break;
      case "rotate-left":
        this.rotate(-90);
        break;
      case "rotate-right":
        this.rotate(90);
        break;
      case "flip-horizontal":
        this.scaleX(-n.scaleX || -1);
        break;
      case "flip-vertical":
        this.scaleY(-n.scaleY || -1);
        break;
      default:
        this.played && this.stop();
    }
  },
  dblclick: function(e) {
    e.preventDefault(), this.viewed && e.target === this.image && ($e && e.isTrusted && clearTimeout(this.doubleClickImageTimeout), this.toggle(e.isTrusted ? e : e.detail && e.detail.originalEvent));
  },
  load: function() {
    var e = this;
    this.timeout && (clearTimeout(this.timeout), this.timeout = !1);
    var i = this.element, n = this.options, r = this.image, a = this.index, s = this.viewerData;
    z(r, He), n.loading && z(this.canvas, Le), r.style.cssText = "height:0;" + "margin-left:".concat(s.width / 2, "px;") + "margin-top:".concat(s.height / 2, "px;") + "max-width:none!important;position:relative;width:0;", this.initImage(function() {
      Xe(r, eo, n.movable), Xe(r, K, n.transition), e.renderImage(function() {
        e.viewed = !0, e.viewing = !1, Y(n.viewed) && C(i, qe, n.viewed, {
          once: !0
        }), G(i, qe, {
          originalImage: e.images[a],
          index: a,
          image: r
        }, {
          cancelable: !1
        });
      });
    });
  },
  loadImage: function(e) {
    var i = e.target, n = i.parentNode, r = n.offsetWidth || 30, a = n.offsetHeight || 50, s = !!Ce(i, "filled");
    fn(i, this.options, function(u, o) {
      var f = u / o, c = r, l = a;
      a * f > r ? s ? c = a * f : l = r / f : s ? l = r / f : c = a * f, se(i, te({
        width: c,
        height: l
      }, at({
        translateX: (r - c) / 2,
        translateY: (a - l) / 2
      })));
    });
  },
  keydown: function(e) {
    var i = this.options;
    if (i.keyboard) {
      var n = e.keyCode || e.which || e.charCode;
      switch (n) {
        case 13:
          this.viewer.contains(e.target) && this.click(e);
          break;
      }
      if (this.fulled)
        switch (n) {
          case 27:
            this.played ? this.stop() : i.inline ? this.fulled && this.exit() : this.hide();
            break;
          case 32:
            this.played && this.stop();
            break;
          case 37:
            this.played && this.playing ? this.playing.prev() : this.prev(i.loop);
            break;
          case 38:
            e.preventDefault(), this.zoom(i.zoomRatio, !0);
            break;
          case 39:
            this.played && this.playing ? this.playing.next() : this.next(i.loop);
            break;
          case 40:
            e.preventDefault(), this.zoom(-i.zoomRatio, !0);
            break;
          case 48:
          case 49:
            e.ctrlKey && (e.preventDefault(), this.toggle());
            break;
        }
    }
  },
  dragstart: function(e) {
    e.target.localName === "img" && e.preventDefault();
  },
  pointerdown: function(e) {
    var i = this.options, n = this.pointers, r = e.buttons, a = e.button;
    if (this.pointerMoved = !1, !(!this.viewed || this.showing || this.viewing || this.hiding || (e.type === "mousedown" || e.type === "pointerdown" && e.pointerType === "mouse") && // No primary button (Usually the left button)
    (B(r) && r !== 1 || B(a) && a !== 0 || e.ctrlKey))) {
      e.preventDefault(), e.changedTouches ? H(e.changedTouches, function(u) {
        n[u.identifier] = tt(u);
      }) : n[e.pointerId || 0] = tt(e);
      var s = i.movable ? nt : !1;
      i.zoomOnTouch && i.zoomable && Object.keys(n).length > 1 ? s = je : i.slideOnTouch && (e.pointerType === "touch" || e.type === "touchstart") && this.isSwitchable() && (s = un), i.transition && (s === nt || s === je) && z(this.image, K), this.action = s;
    }
  },
  pointermove: function(e) {
    var i = this.pointers, n = this.action;
    !this.viewed || !n || (e.preventDefault(), e.changedTouches ? H(e.changedTouches, function(r) {
      te(i[r.identifier] || {}, tt(r, !0));
    }) : te(i[e.pointerId || 0] || {}, tt(e, !0)), this.change(e));
  },
  pointerup: function(e) {
    var i = this, n = this.options, r = this.action, a = this.pointers, s;
    e.changedTouches ? H(e.changedTouches, function(u) {
      s = a[u.identifier], delete a[u.identifier];
    }) : (s = a[e.pointerId || 0], delete a[e.pointerId || 0]), r && (e.preventDefault(), n.transition && (r === nt || r === je) && M(this.image, K), this.action = !1, $e && r !== je && s && Date.now() - s.timeStamp < 500 && (clearTimeout(this.clickCanvasTimeout), clearTimeout(this.doubleClickImageTimeout), n.toggleOnDblclick && this.viewed && e.target === this.image ? this.imageClicked ? (this.imageClicked = !1, this.doubleClickImageTimeout = setTimeout(function() {
      G(i.image, At, {
        originalEvent: e
      });
    }, 50)) : (this.imageClicked = !0, this.doubleClickImageTimeout = setTimeout(function() {
      i.imageClicked = !1;
    }, 500)) : (this.imageClicked = !1, n.backdrop && n.backdrop !== "static" && e.target === this.canvas && (this.clickCanvasTimeout = setTimeout(function() {
      G(i.canvas, Pe, {
        originalEvent: e
      });
    }, 50)))));
  },
  resize: function() {
    var e = this;
    if (!(!this.isShown || this.hiding) && (this.fulled && (this.close(), this.initBody(), this.open()), this.initContainer(), this.initViewer(), this.renderViewer(), this.renderList(), this.viewed && this.initImage(function() {
      e.renderImage();
    }), this.played)) {
      if (this.options.fullscreen && this.fulled && !(document.fullscreenElement || document.webkitFullscreenElement || document.mozFullScreenElement || document.msFullscreenElement)) {
        this.stop();
        return;
      }
      H(this.player.getElementsByTagName("img"), function(i) {
        C(i, re, e.loadImage.bind(e), {
          once: !0
        }), G(i, re);
      });
    }
  },
  wheel: function(e) {
    var i = this;
    if (this.viewed && (e.preventDefault(), !this.wheeling)) {
      this.wheeling = !0, setTimeout(function() {
        i.wheeling = !1;
      }, 50);
      var n = Number(this.options.zoomRatio) || 0.1, r = 1;
      e.deltaY ? r = e.deltaY > 0 ? 1 : -1 : e.wheelDelta ? r = -e.wheelDelta / 120 : e.detail && (r = e.detail > 0 ? 1 : -1), this.zoom(-r * n, !0, null, e);
    }
  }
}, po = {
  /** Show the viewer (only available in modal mode)
   * @param {boolean} [immediate=false] - Indicates if show the viewer immediately or not.
   * @returns {Viewer} this
   */
  show: function() {
    var e = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : !1, i = this.element, n = this.options;
    if (n.inline || this.showing || this.isShown || this.showing)
      return this;
    if (!this.ready)
      return this.build(), this.ready && this.show(e), this;
    if (Y(n.show) && C(i, Ei, n.show, {
      once: !0
    }), G(i, Ei) === !1 || !this.ready)
      return this;
    this.hiding && this.transitioning.abort(), this.showing = !0, this.open();
    var r = this.viewer;
    if (z(r, _e), r.setAttribute("role", "dialog"), r.setAttribute("aria-labelledby", this.title.id), r.setAttribute("aria-modal", !0), r.removeAttribute("aria-hidden"), n.transition && !e) {
      var a = this.shown.bind(this);
      this.transitioning = {
        abort: function() {
          j(r, ae, a), z(r, ne);
        }
      }, M(r, K), r.initialOffsetWidth = r.offsetWidth, C(r, ae, a, {
        once: !0
      }), M(r, ne);
    } else
      M(r, ne), this.shown();
    return this;
  },
  /**
   * Hide the viewer (only available in modal mode)
   * @param {boolean} [immediate=false] - Indicates if hide the viewer immediately or not.
   * @returns {Viewer} this
   */
  hide: function() {
    var e = this, i = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : !1, n = this.element, r = this.options;
    if (r.inline || this.hiding || !(this.isShown || this.showing))
      return this;
    if (Y(r.hide) && C(n, Di, r.hide, {
      once: !0
    }), G(n, Di) === !1)
      return this;
    this.showing && this.transitioning.abort(), this.hiding = !0, this.played ? this.stop() : this.viewing && this.viewing.abort();
    var a = this.viewer, s = this.image, u = function() {
      z(a, ne), e.hidden();
    };
    if (r.transition && !i) {
      var o = function(l) {
        l && l.target === a && (j(a, ae, o), e.hidden());
      }, f = function() {
        Ae(a, K) ? (C(a, ae, o), z(a, ne)) : u();
      };
      this.transitioning = {
        abort: function() {
          e.viewed && Ae(s, K) ? j(s, ae, f) : Ae(a, K) && j(a, ae, o);
        }
      }, this.viewed && Ae(s, K) ? (C(s, ae, f, {
        once: !0
      }), this.zoomTo(0, !1, null, null, !0)) : f();
    } else
      u();
    return this;
  },
  /**
   * View one of the images with image's index
   * @param {number} index - The index of the image to view.
   * @returns {Viewer} this
   */
  view: function() {
    var e = this, i = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : this.options.initialViewIndex;
    if (i = Number(i) || 0, this.hiding || this.played || i < 0 || i >= this.length || this.viewed && i === this.index)
      return this;
    if (!this.isShown)
      return this.index = i, this.show();
    this.viewing && this.viewing.abort();
    var n = this.element, r = this.options, a = this.title, s = this.canvas, u = this.items[i], o = u.querySelector("img"), f = Ce(o, "originalUrl"), c = o.getAttribute("alt"), l = document.createElement("img");
    if (H(r.inheritedAttributes, function(g) {
      var m = o.getAttribute(g);
      m !== null && l.setAttribute(g, m);
    }), l.src = f, l.alt = c, Y(r.view) && C(n, Oi, r.view, {
      once: !0
    }), G(n, Oi, {
      originalImage: this.images[i],
      index: i,
      image: l
    }) === !1 || !this.isShown || this.hiding || this.played)
      return this;
    var d = this.items[this.index];
    d && (z(d, Je), d.removeAttribute("aria-selected")), M(u, Je), u.setAttribute("aria-selected", !0), r.focus && u.focus(), this.image = l, this.viewed = !1, this.index = i, this.imageData = {}, M(l, He), r.loading && M(s, Le), s.innerHTML = "", s.appendChild(l), this.renderList(), a.innerHTML = "";
    var b = function() {
      var m = e.imageData, _ = Array.isArray(r.title) ? r.title[1] : r.title;
      a.innerHTML = oo(Y(_) ? _.call(e, l, m) : "".concat(c, " (").concat(m.naturalWidth, " × ").concat(m.naturalHeight, ")"));
    }, h, E;
    return C(n, qe, b, {
      once: !0
    }), this.viewing = {
      abort: function() {
        j(n, qe, b), l.complete ? e.imageRendering ? e.imageRendering.abort() : e.imageInitializing && e.imageInitializing.abort() : (l.src = "", j(l, re, h), e.timeout && clearTimeout(e.timeout));
      }
    }, l.complete ? this.load() : (C(l, re, h = function() {
      j(l, Se, E), e.load();
    }, {
      once: !0
    }), C(l, Se, E = function() {
      j(l, re, h), e.timeout && (clearTimeout(e.timeout), e.timeout = !1), z(l, He), r.loading && z(e.canvas, Le);
    }, {
      once: !0
    }), this.timeout && clearTimeout(this.timeout), this.timeout = setTimeout(function() {
      z(l, He), e.timeout = !1;
    }, 1e3)), this;
  },
  /**
   * View the previous image
   * @param {boolean} [loop=false] - Indicate if view the last one
   * when it is the first one at present.
   * @returns {Viewer} this
   */
  prev: function() {
    var e = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : !1, i = this.index - 1;
    return i < 0 && (i = e ? this.length - 1 : 0), this.view(i), this;
  },
  /**
   * View the next image
   * @param {boolean} [loop=false] - Indicate if view the first one
   * when it is the last one at present.
   * @returns {Viewer} this
   */
  next: function() {
    var e = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : !1, i = this.length - 1, n = this.index + 1;
    return n > i && (n = e ? 0 : i), this.view(n), this;
  },
  /**
   * Move the image with relative offsets.
   * @param {number} x - The moving distance in the horizontal direction.
   * @param {number} [y=x] The moving distance in the vertical direction.
   * @returns {Viewer} this
   */
  move: function(e) {
    var i = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : e, n = this.imageData;
    return this.moveTo(xe(e) ? e : n.x + Number(e), xe(i) ? i : n.y + Number(i)), this;
  },
  /**
   * Move the image to an absolute point.
   * @param {number} x - The new position in the horizontal direction.
   * @param {number} [y=x] - The new position in the vertical direction.
   * @param {Event} [_originalEvent=null] - The original event if any.
   * @returns {Viewer} this
   */
  moveTo: function(e) {
    var i = this, n = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : e, r = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : null, a = this.element, s = this.options, u = this.imageData;
    if (e = Number(e), n = Number(n), this.viewed && !this.played && s.movable) {
      var o = u.x, f = u.y, c = !1;
      if (B(e) ? c = !0 : e = o, B(n) ? c = !0 : n = f, c) {
        if (Y(s.move) && C(a, xi, s.move, {
          once: !0
        }), G(a, xi, {
          x: e,
          y: n,
          oldX: o,
          oldY: f,
          originalEvent: r
        }) === !1)
          return this;
        u.x = e, u.y = n, u.left = e, u.top = n, this.moving = !0, this.renderImage(function() {
          i.moving = !1, Y(s.moved) && C(a, Mi, s.moved, {
            once: !0
          }), G(a, Mi, {
            x: e,
            y: n,
            oldX: o,
            oldY: f,
            originalEvent: r
          }, {
            cancelable: !1
          });
        });
      }
    }
    return this;
  },
  /**
   * Rotate the image with a relative degree.
   * @param {number} degree - The rotate degree.
   * @returns {Viewer} this
   */
  rotate: function(e) {
    return this.rotateTo((this.imageData.rotate || 0) + Number(e)), this;
  },
  /**
   * Rotate the image to an absolute degree.
   * @param {number} degree - The rotate degree.
   * @returns {Viewer} this
   */
  rotateTo: function(e) {
    var i = this, n = this.element, r = this.options, a = this.imageData;
    if (e = Number(e), B(e) && this.viewed && !this.played && r.rotatable) {
      var s = a.rotate;
      if (Y(r.rotate) && C(n, Ci, r.rotate, {
        once: !0
      }), G(n, Ci, {
        degree: e,
        oldDegree: s
      }) === !1)
        return this;
      a.rotate = e, this.rotating = !0, this.renderImage(function() {
        i.rotating = !1, Y(r.rotated) && C(n, Ii, r.rotated, {
          once: !0
        }), G(n, Ii, {
          degree: e,
          oldDegree: s
        }, {
          cancelable: !1
        });
      });
    }
    return this;
  },
  /**
   * Scale the image on the x-axis.
   * @param {number} scaleX - The scale ratio on the x-axis.
   * @returns {Viewer} this
   */
  scaleX: function(e) {
    return this.scale(e, this.imageData.scaleY), this;
  },
  /**
   * Scale the image on the y-axis.
   * @param {number} scaleY - The scale ratio on the y-axis.
   * @returns {Viewer} this
   */
  scaleY: function(e) {
    return this.scale(this.imageData.scaleX, e), this;
  },
  /**
   * Scale the image.
   * @param {number} scaleX - The scale ratio on the x-axis.
   * @param {number} [scaleY=scaleX] - The scale ratio on the y-axis.
   * @returns {Viewer} this
   */
  scale: function(e) {
    var i = this, n = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : e, r = this.element, a = this.options, s = this.imageData;
    if (e = Number(e), n = Number(n), this.viewed && !this.played && a.scalable) {
      var u = s.scaleX, o = s.scaleY, f = !1;
      if (B(e) ? f = !0 : e = u, B(n) ? f = !0 : n = o, f) {
        if (Y(a.scale) && C(r, $i, a.scale, {
          once: !0
        }), G(r, $i, {
          scaleX: e,
          scaleY: n,
          oldScaleX: u,
          oldScaleY: o
        }) === !1)
          return this;
        s.scaleX = e, s.scaleY = n, this.scaling = !0, this.renderImage(function() {
          i.scaling = !1, Y(a.scaled) && C(r, Li, a.scaled, {
            once: !0
          }), G(r, Li, {
            scaleX: e,
            scaleY: n,
            oldScaleX: u,
            oldScaleY: o
          }, {
            cancelable: !1
          });
        });
      }
    }
    return this;
  },
  /**
   * Zoom the image with a relative ratio.
   * @param {number} ratio - The target ratio.
   * @param {boolean} [showTooltip=false] - Indicates whether to show the tooltip.
   * @param {Object} [pivot] - The pivot point coordinate for zooming.
   * @param {Event} [_originalEvent=null] - The original event if any.
   * @returns {Viewer} this
   */
  zoom: function(e) {
    var i = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : !1, n = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : null, r = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : null, a = this.imageData;
    return e = Number(e), e < 0 ? e = 1 / (1 - e) : e = 1 + e, this.zoomTo(a.width * e / a.naturalWidth, i, n, r), this;
  },
  /**
   * Zoom the image to an absolute ratio.
   * @param {number} ratio - The target ratio.
   * @param {boolean} [showTooltip] - Indicates whether to show the tooltip.
   * @param {Object} [pivot] - The pivot point coordinate for zooming.
   * @param {Event} [_originalEvent=null] - The original event if any.
   * @param {Event} [_zoomable=false] - Indicates if the current zoom is available or not.
   * @returns {Viewer} this
   */
  zoomTo: function(e) {
    var i = this, n = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : !1, r = arguments.length > 2 && arguments[2] !== void 0 ? arguments[2] : null, a = arguments.length > 3 && arguments[3] !== void 0 ? arguments[3] : null, s = arguments.length > 4 && arguments[4] !== void 0 ? arguments[4] : !1, u = this.element, o = this.options, f = this.pointers, c = this.imageData, l = c.x, d = c.y, b = c.width, h = c.height, E = c.naturalWidth, g = c.naturalHeight;
    if (e = Math.max(0, e), B(e) && this.viewed && !this.played && (s || o.zoomable)) {
      if (!s) {
        var m = Math.max(0.01, o.minZoomRatio), _ = Math.min(100, o.maxZoomRatio);
        e = Math.min(Math.max(e, m), _);
      }
      if (a)
        switch (a.type) {
          case "wheel":
            o.zoomRatio >= 0.055 && e > 0.95 && e < 1.05 && (e = 1);
            break;
          case "pointermove":
          case "touchmove":
          case "mousemove":
            e > 0.99 && e < 1.01 && (e = 1);
            break;
        }
      var S = E * e, P = g * e, N = S - b, F = P - h, X = c.ratio;
      if (Y(o.zoom) && C(u, Ri, o.zoom, {
        once: !0
      }), G(u, Ri, {
        ratio: e,
        oldRatio: X,
        originalEvent: a
      }) === !1)
        return this;
      if (this.zooming = !0, a) {
        var W = uo(this.viewer), R = f && Object.keys(f).length > 0 ? ho(f) : {
          pageX: a.pageX,
          pageY: a.pageY
        };
        c.x -= N * ((R.pageX - W.left - l) / b), c.y -= F * ((R.pageY - W.top - d) / h);
      } else Me(r) && B(r.x) && B(r.y) ? (c.x -= N * ((r.x - l) / b), c.y -= F * ((r.y - d) / h)) : (c.x -= N / 2, c.y -= F / 2);
      c.left = c.x, c.top = c.y, c.width = S, c.height = P, c.oldRatio = X, c.ratio = e, this.renderImage(function() {
        i.zooming = !1, Y(o.zoomed) && C(u, Pi, o.zoomed, {
          once: !0
        }), G(u, Pi, {
          ratio: e,
          oldRatio: X,
          originalEvent: a
        }, {
          cancelable: !1
        });
      }), n && this.tooltip();
    }
    return this;
  },
  /**
   * Play the images
   * @param {boolean|FullscreenOptions} [fullscreen=false] - Indicate if request fullscreen or not.
   * @returns {Viewer} this
   */
  play: function() {
    var e = this, i = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : !1;
    if (!this.isShown || this.played)
      return this;
    var n = this.element, r = this.options;
    if (Y(r.play) && C(n, Ni, r.play, {
      once: !0
    }), G(n, Ni) === !1)
      return this;
    var a = this.player, s = this.loadImage.bind(this), u = [], o = 0, f = 0;
    if (this.played = !0, this.onLoadWhenPlay = s, i && this.requestFullscreen(i), M(a, De), H(this.items, function(d, b) {
      var h = d.querySelector("img"), E = document.createElement("img");
      E.src = Ce(h, "originalUrl"), E.alt = h.getAttribute("alt"), E.referrerPolicy = h.referrerPolicy, o += 1, M(E, rt), Xe(E, K, r.transition), Ae(d, Je) && (M(E, ne), f = b), u.push(E), C(E, re, s, {
        once: !0
      }), a.appendChild(E);
    }), B(r.interval) && r.interval > 0) {
      var c = function() {
        clearTimeout(e.playing.timeout), z(u[f], ne), f -= 1, f = f >= 0 ? f : o - 1, M(u[f], ne), e.playing.timeout = setTimeout(c, r.interval);
      }, l = function() {
        clearTimeout(e.playing.timeout), z(u[f], ne), f += 1, f = f < o ? f : 0, M(u[f], ne), e.playing.timeout = setTimeout(l, r.interval);
      };
      o > 1 && (this.playing = {
        prev: c,
        next: l,
        timeout: setTimeout(l, r.interval)
      });
    }
    return this;
  },
  // Stop play
  stop: function() {
    var e = this;
    if (!this.played)
      return this;
    var i = this.element, n = this.options;
    if (Y(n.stop) && C(i, zi, n.stop, {
      once: !0
    }), G(i, zi) === !1)
      return this;
    var r = this.player;
    return clearTimeout(this.playing.timeout), this.playing = !1, this.played = !1, H(r.getElementsByTagName("img"), function(a) {
      j(a, re, e.onLoadWhenPlay);
    }), z(r, De), r.innerHTML = "", this.exitFullscreen(), this;
  },
  // Enter modal mode (only available in inline mode)
  full: function() {
    var e = this, i = this.options, n = this.viewer, r = this.image, a = this.list;
    return !this.isShown || this.played || this.fulled || !i.inline ? this : (this.fulled = !0, this.open(), M(this.button, hi), i.transition && (z(a, K), this.viewed && z(r, K)), M(n, Dt), n.setAttribute("role", "dialog"), n.setAttribute("aria-labelledby", this.title.id), n.setAttribute("aria-modal", !0), n.removeAttribute("style"), se(n, {
      zIndex: i.zIndex
    }), i.focus && this.enforceFocus(), this.initContainer(), this.viewerData = te({}, this.containerData), this.renderList(), this.viewed && this.initImage(function() {
      e.renderImage(function() {
        i.transition && setTimeout(function() {
          M(r, K), M(a, K);
        }, 0);
      });
    }), this);
  },
  // Exit modal mode (only available in inline mode)
  exit: function() {
    var e = this, i = this.options, n = this.viewer, r = this.image, a = this.list;
    return !this.isShown || this.played || !this.fulled || !i.inline ? this : (this.fulled = !1, this.close(), z(this.button, hi), i.transition && (z(a, K), this.viewed && z(r, K)), i.focus && this.clearEnforceFocus(), n.removeAttribute("role"), n.removeAttribute("aria-labelledby"), n.removeAttribute("aria-modal"), z(n, Dt), se(n, {
      zIndex: i.zIndexInline
    }), this.viewerData = te({}, this.parentData), this.renderViewer(), this.renderList(), this.viewed && this.initImage(function() {
      e.renderImage(function() {
        i.transition && setTimeout(function() {
          M(r, K), M(a, K);
        }, 0);
      });
    }), this);
  },
  // Show the current ratio of the image with percentage
  tooltip: function() {
    var e = this, i = this.options, n = this.tooltipBox, r = this.imageData;
    return !this.viewed || this.played || !i.tooltip ? this : (n.textContent = "".concat(Math.round(r.ratio * 100), "%"), this.tooltipping ? clearTimeout(this.tooltipping) : i.transition ? (this.fading && G(n, ae), M(n, De), M(n, rt), M(n, K), n.removeAttribute("aria-hidden"), n.initialOffsetWidth = n.offsetWidth, M(n, ne)) : (M(n, De), n.removeAttribute("aria-hidden")), this.tooltipping = setTimeout(function() {
      i.transition ? (C(n, ae, function() {
        z(n, De), z(n, rt), z(n, K), n.setAttribute("aria-hidden", !0), e.fading = !1;
      }, {
        once: !0
      }), z(n, ne), e.fading = !0) : (z(n, De), n.setAttribute("aria-hidden", !0)), e.tooltipping = !1;
    }, 1e3), this);
  },
  /**
   * Toggle the image size between its current size and natural size
   * @param {Event} [_originalEvent=null] - The original event if any.
   * @returns {Viewer} this
   */
  toggle: function() {
    var e = arguments.length > 0 && arguments[0] !== void 0 ? arguments[0] : null;
    return this.imageData.ratio === 1 ? this.zoomTo(this.imageData.oldRatio, !0, null, e) : this.zoomTo(1, !0, null, e), this;
  },
  // Reset the image to its initial state
  reset: function() {
    return this.viewed && !this.played && (this.imageData = te({}, this.initialImageData), this.renderImage()), this;
  },
  // Update viewer when images changed
  update: function() {
    var e = this, i = this.element, n = this.options, r = this.isImg;
    if (r && !i.parentNode)
      return this.destroy();
    var a = [];
    if (H(r ? [i] : i.querySelectorAll("img"), function(f) {
      Y(n.filter) ? n.filter.call(e, f) && a.push(f) : e.getImageURL(f) && a.push(f);
    }), !a.length)
      return this;
    if (this.images = a, this.length = a.length, this.ready) {
      var s = [];
      if (H(this.items, function(f, c) {
        var l = f.querySelector("img"), d = a[c];
        d && l ? (d.src !== l.src || d.alt !== l.alt) && s.push(c) : s.push(c);
      }), se(this.list, {
        width: "auto"
      }), this.initList(), this.isShown)
        if (this.length) {
          if (this.viewed) {
            var u = s.indexOf(this.index);
            if (u >= 0)
              this.viewed = !1, this.view(Math.max(Math.min(this.index - u, this.length - 1), 0));
            else {
              var o = this.items[this.index];
              M(o, Je), o.setAttribute("aria-selected", !0);
            }
          }
        } else
          this.image = null, this.viewed = !1, this.index = 0, this.imageData = {}, this.canvas.innerHTML = "", this.title.innerHTML = "";
    } else
      this.build();
    return this;
  },
  // Destroy the viewer
  destroy: function() {
    var e = this.element, i = this.options;
    return e[L] ? (this.destroyed = !0, this.ready ? (this.played && this.stop(), i.inline ? (this.fulled && this.exit(), this.unbind()) : this.isShown ? (this.viewing && (this.imageRendering ? this.imageRendering.abort() : this.imageInitializing && this.imageInitializing.abort()), this.hiding && this.transitioning.abort(), this.hidden()) : this.showing && (this.transitioning.abort(), this.hidden()), this.ready = !1, this.viewer.parentNode.removeChild(this.viewer)) : i.inline && (this.delaying ? this.delaying.abort() : this.initializing && this.initializing.abort()), i.inline || j(e, Pe, this.onStart), e[L] = void 0, this) : this;
  }
}, wo = {
  getImageURL: function(e) {
    var i = this.options.url;
    return Be(i) ? i = e.getAttribute(i) : Y(i) ? i = i.call(this, e) : i = "", i;
  },
  enforceFocus: function() {
    var e = this;
    this.clearEnforceFocus(), C(document, gi, this.onFocusin = function(i) {
      var n = e.viewer, r = i.target;
      if (!(r === document || r === n || n.contains(r))) {
        for (; r; ) {
          if (r.getAttribute("tabindex") !== null || r.getAttribute("aria-modal") === "true")
            return;
          r = r.parentElement;
        }
        n.focus();
      }
    });
  },
  clearEnforceFocus: function() {
    this.onFocusin && (j(document, gi, this.onFocusin), this.onFocusin = null);
  },
  open: function() {
    var e = this.body;
    M(e, di), this.scrollbarWidth > 0 && (e.style.paddingRight = "".concat(this.scrollbarWidth + (parseFloat(this.initialBodyComputedPaddingRight) || 0), "px"));
  },
  close: function() {
    var e = this.body;
    z(e, di), this.scrollbarWidth > 0 && (e.style.paddingRight = this.initialBodyPaddingRight);
  },
  shown: function() {
    var e = this.element, i = this.options, n = this.viewer;
    this.fulled = !0, this.isShown = !0, this.render(), this.bind(), this.showing = !1, i.focus && (n.focus(), this.enforceFocus()), Y(i.shown) && C(e, Ti, i.shown, {
      once: !0
    }), G(e, Ti) !== !1 && this.ready && this.isShown && !this.hiding && this.view(this.index);
  },
  hidden: function() {
    var e = this.element, i = this.options, n = this.viewer;
    i.fucus && this.clearEnforceFocus(), this.close(), this.unbind(), M(n, _e), n.removeAttribute("role"), n.removeAttribute("aria-labelledby"), n.removeAttribute("aria-modal"), n.setAttribute("aria-hidden", !0), this.resetList(), this.resetImage(), this.fulled = !1, this.viewed = !1, this.isShown = !1, this.hiding = !1, this.destroyed || (Y(i.hidden) && C(e, Ai, i.hidden, {
      once: !0
    }), G(e, Ai, null, {
      cancelable: !1
    }));
  },
  requestFullscreen: function(e) {
    var i = this.element.ownerDocument;
    if (this.fulled && !(i.fullscreenElement || i.webkitFullscreenElement || i.mozFullScreenElement || i.msFullscreenElement)) {
      var n = i.documentElement;
      n.requestFullscreen ? Me(e) ? n.requestFullscreen(e) : n.requestFullscreen() : n.webkitRequestFullscreen ? n.webkitRequestFullscreen(Element.ALLOW_KEYBOARD_INPUT) : n.mozRequestFullScreen ? n.mozRequestFullScreen() : n.msRequestFullscreen && n.msRequestFullscreen();
    }
  },
  exitFullscreen: function() {
    var e = this.element.ownerDocument;
    this.fulled && (e.fullscreenElement || e.webkitFullscreenElement || e.mozFullScreenElement || e.msFullscreenElement) && (e.exitFullscreen ? e.exitFullscreen() : e.webkitExitFullscreen ? e.webkitExitFullscreen() : e.mozCancelFullScreen ? e.mozCancelFullScreen() : e.msExitFullscreen && e.msExitFullscreen());
  },
  change: function(e) {
    var i = this.options, n = this.pointers, r = n[Object.keys(n)[0]];
    if (r) {
      var a = r.endX - r.startX, s = r.endY - r.startY;
      switch (this.action) {
        case nt:
          (a !== 0 || s !== 0) && (this.pointerMoved = !0, this.move(a, s, e));
          break;
        case je:
          this.zoom(fo(n), !1, null, e);
          break;
        case un: {
          this.action = "switched";
          var u = Math.abs(a);
          u > 1 && u > Math.abs(s) && (this.pointers = {}, a > 1 ? this.prev(i.loop) : a < -1 && this.next(i.loop));
          break;
        }
      }
      H(n, function(o) {
        o.startX = o.endX, o.startY = o.endY;
      });
    }
  },
  isSwitchable: function() {
    var e = this.imageData, i = this.viewerData;
    return this.length > 1 && e.x >= 0 && e.y >= 0 && e.width <= i.width && e.height <= i.height;
  }
}, yo = he.Viewer, bo = /* @__PURE__ */ function(t) {
  return function() {
    return t += 1, t;
  };
}(-1), hn = /* @__PURE__ */ function() {
  function t(e) {
    var i = arguments.length > 1 && arguments[1] !== void 0 ? arguments[1] : {};
    if (Vs(this, t), !e || e.nodeType !== 1)
      throw new Error("The first argument is required and must be an element.");
    this.element = e, this.options = te({}, fi, Me(i) && i), this.action = !1, this.fading = !1, this.fulled = !1, this.hiding = !1, this.imageClicked = !1, this.imageData = {}, this.index = this.options.initialViewIndex, this.isImg = !1, this.isShown = !1, this.length = 0, this.moving = !1, this.played = !1, this.playing = !1, this.pointers = {}, this.ready = !1, this.rotating = !1, this.scaling = !1, this.showing = !1, this.timeout = !1, this.tooltipping = !1, this.viewed = !1, this.viewing = !1, this.wheeling = !1, this.zooming = !1, this.pointerMoved = !1, this.id = bo(), this.init();
  }
  return Ws(t, [{
    key: "init",
    value: function() {
      var i = this, n = this.element, r = this.options;
      if (!n[L]) {
        n[L] = this, r.focus && !r.keyboard && (r.focus = !1);
        var a = n.localName === "img", s = [];
        if (H(a ? [n] : n.querySelectorAll("img"), function(f) {
          Y(r.filter) ? r.filter.call(i, f) && s.push(f) : i.getImageURL(f) && s.push(f);
        }), this.isImg = a, this.length = s.length, this.images = s, this.initBody(), xe(document.createElement(L).style.transition) && (r.transition = !1), r.inline) {
          var u = 0, o = function() {
            if (u += 1, u === i.length) {
              var c;
              i.initializing = !1, i.delaying = {
                abort: function() {
                  clearTimeout(c);
                }
              }, c = setTimeout(function() {
                i.delaying = !1, i.build();
              }, 0);
            }
          };
          this.initializing = {
            abort: function() {
              H(s, function(c) {
                c.complete || (j(c, re, o), j(c, Se, o));
              });
            }
          }, H(s, function(f) {
            if (f.complete)
              o();
            else {
              var c, l;
              C(f, re, c = function() {
                j(f, Se, l), o();
              }, {
                once: !0
              }), C(f, Se, l = function() {
                j(f, re, c), o();
              }, {
                once: !0
              });
            }
          });
        } else
          C(n, Pe, this.onStart = function(f) {
            var c = f.target;
            c.localName === "img" && (!Y(r.filter) || r.filter.call(i, c)) && i.view(i.images.indexOf(c));
          });
      }
    }
  }, {
    key: "build",
    value: function() {
      if (!this.ready) {
        var i = this.element, n = this.options, r = i.parentNode, a = document.createElement("div");
        a.innerHTML = Xs;
        var s = a.querySelector(".".concat(L, "-container")), u = s.querySelector(".".concat(L, "-title")), o = s.querySelector(".".concat(L, "-toolbar")), f = s.querySelector(".".concat(L, "-navbar")), c = s.querySelector(".".concat(L, "-button")), l = s.querySelector(".".concat(L, "-canvas"));
        if (this.parent = r, this.viewer = s, this.title = u, this.toolbar = o, this.navbar = f, this.button = c, this.canvas = l, this.footer = s.querySelector(".".concat(L, "-footer")), this.tooltipBox = s.querySelector(".".concat(L, "-tooltip")), this.player = s.querySelector(".".concat(L, "-player")), this.list = s.querySelector(".".concat(L, "-list")), s.id = "".concat(L).concat(this.id), u.id = "".concat(L, "Title").concat(this.id), M(u, n.title ? et(Array.isArray(n.title) ? n.title[0] : n.title) : _e), M(f, n.navbar ? et(n.navbar) : _e), Xe(c, _e, !n.button), n.keyboard && c.setAttribute("tabindex", 0), n.backdrop && (M(s, "".concat(L, "-backdrop")), !n.inline && n.backdrop !== "static" && Ot(l, st, "hide")), Be(n.className) && n.className && n.className.split(Yt).forEach(function(S) {
          M(s, S);
        }), n.toolbar) {
          var d = document.createElement("ul"), b = Me(n.toolbar), h = Qe.slice(0, 3), E = Qe.slice(7, 9), g = Qe.slice(9);
          b || M(o, et(n.toolbar)), H(b ? n.toolbar : Qe, function(S, P) {
            var N = b && Me(S), F = b ? jt(P) : S, X = N && !xe(S.show) ? S.show : S;
            if (!(!X || !n.zoomable && h.indexOf(F) !== -1 || !n.rotatable && E.indexOf(F) !== -1 || !n.scalable && g.indexOf(F) !== -1)) {
              var W = N && !xe(S.size) ? S.size : S, R = N && !xe(S.click) ? S.click : S, A = document.createElement("li");
              n.keyboard && A.setAttribute("tabindex", 0), A.setAttribute("role", "button"), M(A, "".concat(L, "-").concat(F)), Y(R) || Ot(A, st, F), B(X) && M(A, et(X)), ["small", "large"].indexOf(W) !== -1 ? M(A, "".concat(L, "-").concat(W)) : F === "play" && M(A, "".concat(L, "-large")), Y(R) && C(A, Pe, R), d.appendChild(A);
            }
          }), o.appendChild(d);
        } else
          M(o, _e);
        if (!n.rotatable) {
          var m = o.querySelectorAll('li[class*="rotate"]');
          M(m, He), H(m, function(S) {
            o.appendChild(S);
          });
        }
        if (n.inline)
          M(c, Gs), se(s, {
            zIndex: n.zIndexInline
          }), window.getComputedStyle(r).position === "static" && se(r, {
            position: "relative"
          }), r.insertBefore(s, i.nextSibling);
        else {
          M(c, Ks), M(s, Dt), M(s, rt), M(s, _e), se(s, {
            zIndex: n.zIndex
          });
          var _ = n.container;
          Be(_) && (_ = i.ownerDocument.querySelector(_)), _ || (_ = this.body), _.appendChild(s);
        }
        if (n.inline && (this.render(), this.bind(), this.isShown = !0), this.ready = !0, Y(n.ready) && C(i, Si, n.ready, {
          once: !0
        }), G(i, Si) === !1) {
          this.ready = !1;
          return;
        }
        this.ready && n.inline && this.view(this.index);
      }
    }
    /**
     * Get the no conflict viewer class.
     * @returns {Viewer} The viewer class.
     */
  }], [{
    key: "noConflict",
    value: function() {
      return window.Viewer = yo, t;
    }
    /**
     * Change the default options.
     * @param {Object} options - The new default options.
     */
  }, {
    key: "setDefaults",
    value: function(i) {
      te(fi, Me(i) && i);
    }
  }]);
}();
te(hn.prototype, vo, go, mo, po, wo);
function Ro(t) {
  const e = Q([]), i = Q(!1), n = xt("speed-components-config", Q({ apis: {} })), r = Oe(() => {
    var l, d, b, h;
    return t === void 0 ? (l = n.value) != null && l.apis ? { apis: (d = n.value) == null ? void 0 : d.apis } : {} : (b = n.value) != null && b.apis ? { apis: (h = n.value) == null ? void 0 : h.apis, ...t.value } : t.value;
  }), a = (l, d) => {
    var m, _;
    const b = d || e.value;
    if (!r.value)
      return !0;
    let h = !1, E = !0, g = !0;
    if (r.value.acceptTypes && r.value.acceptTypes.length > 0) {
      const S = /\.\w+$/, P = ((m = l.name.match(S)) == null ? void 0 : m[0]) ?? "";
      r.value.acceptTypes.includes(P) || (it.warning(
        `请上传${r.value.acceptTypes.join("、")}格式的文件`
      ), E = !1);
    }
    return r.value.maxCount && r.value.maxCount > 1 && (h = e.value.length + ((_ = b || []) == null ? void 0 : _.length) > r.value.maxCount, h) ? (it.warning(`已超过最大上传数${r.value.maxCount}`), !1) : (r.value.maxSize && l.size && (g = l.size / 1024 / 1024 < r.value.maxSize, g || it.warning(`单个文件最大为${r.value.maxSize}M`)), g && E);
  }, s = async (l) => {
    var _, S, P, N, F, X, W;
    if (r.value.multiple && !((S = (_ = r == null ? void 0 : r.value) == null ? void 0 : _.apis) != null && S.fileUploadMulti)) {
      console.warn("未配置批量上传接口");
      return;
    } else if (!((N = (P = r == null ? void 0 : r.value) == null ? void 0 : P.apis) != null && N.fileUploadSingle)) {
      console.warn("未配置单条上传接口");
      return;
    }
    let d = new FormData();
    const { file: b } = l;
    let h = [];
    if (r.value.multiple && Array.isArray(b) ? (h = b, b.forEach((R) => {
      d.append(r.value.name || "files[]", R);
    })) : (h = [b], d.append(r.value.name || "file", b)), (F = r.value) != null && F.data) {
      if (!(typeof r.value.data == "function" || typeof r.value.data == "object")) {
        console.warn("data类型错误,请传入object或function");
        return;
      }
      if (typeof r.value.data == "object") {
        const R = r.value.data;
        Object.keys(R).forEach((A) => {
          d.append(A, R[A]);
        });
      } else {
        const R = r.value.data(
          h
        );
        Array.from(R.entries()).forEach(([A, p]) => {
          d.append(A, p);
        });
      }
    }
    console.log("formData-imgUpload=", d), i.value = !0;
    const E = h.map(() => ks()), g = h.map((R, A) => ({
      uid: E[A],
      fileName: R.name,
      status: "uploading"
    }));
    r.value.maxCount === 1 ? e.value = [g[0]] : e.value.push(...g);
    const m = r.value.multiple ? r.value.apis.fileUploadMulti : r.value.apis.fileUploadSingle;
    try {
      const R = await m(d);
      i.value = !1, R.success ? ((r.value.multiple ? R.data : [R.data]).map(
        (p) => {
          var I;
          return (I = r == null ? void 0 : r.value) != null && I.transformResult && typeof r.value.transformResult == "function" ? r.value.transformResult(p) : {
            id: p.id,
            fileType: p.fileType,
            fileSize: p.fileSize,
            fileName: p.fileName
          };
        }
      ).forEach((p, I) => {
        const k = e.value.findIndex(
          (T) => T.uid === E[I]
        );
        k !== -1 && (e.value[k] = { ...p, status: "done" });
      }), (X = l.afterUpload) == null || X.call(l, e.value)) : E.forEach((A) => {
        const p = e.value.findIndex(
          (I) => I.uid === A
        );
        p !== -1 && (e.value[p].status = "error");
      }), (W = r == null ? void 0 : r.value) != null && W.afterUpload && typeof r.value.afterUpload == "function" && r.value.afterUpload(e.value);
    } catch {
      i.value = !1, E.forEach((A) => {
        const p = e.value.findIndex(
          (I) => I.uid === A
        );
        p !== -1 && (e.value[p].status = "error");
      });
    }
  }, u = Q();
  return {
    beforeUpload: a,
    customRequest: s,
    handleDelFile: async (l) => {
      var h, E, g, m, _;
      const d = l.id;
      if (!((h = r == null ? void 0 : r.value) != null && h.apis)) {
        console.warn("未配置删除接口");
        const S = e.value.findIndex((P) => P.id === d);
        e.value.splice(S, 1), (E = r == null ? void 0 : r.value) != null && E.afterDelete && typeof r.value.afterDelete == "function" && r.value.afterDelete(e.value);
        return;
      }
      i.value = !0;
      const b = await ((m = (g = r.value.apis).fileDel) == null ? void 0 : m.call(g, d));
      if (i.value = !1, b && b.success) {
        const S = e.value.findIndex((P) => P.id === d);
        e.value.splice(S, 1), (_ = r == null ? void 0 : r.value) != null && _.afterDelete && typeof r.value.afterDelete == "function" && r.value.afterDelete(e.value);
      }
    },
    handleDownloadFile: async (l) => {
      var h, E;
      const d = (E = (h = r.value) == null ? void 0 : h.apis) == null ? void 0 : E.fileDownload;
      if (!d || typeof d != "function")
        throw new Error("文件下载方法缺失或传入的fileDownload不正确");
      const b = await d(l.id);
      Us(b, l.fileName, l == null ? void 0 : l.className);
    },
    handlePreviewFile: async (l) => {
      var E, g, m;
      if (!l.fileName)
        throw new Error("缺少必要属性fileName");
      const d = (g = (E = r.value) == null ? void 0 : E.apis) == null ? void 0 : g.getPreviewUrl, b = /\.\w+$/, h = ((m = l.fileName.match(b)) == null ? void 0 : m[0]) ?? "";
      if (h)
        if ([".png", ".jpg", ".gif", ".jpeg"].includes(h))
          if (u.value)
            u.value.view(0);
          else {
            let _ = null;
            document.getElementById("viewer-img") ? _ = document.getElementById("viewer-img") : (_ = document.createElement("img"), _.id = "viewer-img", _.style.display = "none"), _ && (_.src = (d ? d(l.id) : l.previewUrl) || "", u.value = new hn(_, {
              className: l.className,
              // 内联展示
              inline: !1
            }), u.value.view(0));
          }
        else
          r.value.onPreview || typeof r.value.onPreview == "function" ? r.value.onPreview(l.id) : window.open(l.previewUrl);
      else
        r.value.onPreview || typeof r.value.onPreview == "function" ? r.value.onPreview(l.id) : window.open(l.previewUrl);
    },
    uploadLoading: i,
    files: e
  };
}
const _o = {
  extraParams: {},
  needFullSelect: !1,
  hasPagination: !0,
  hasSort: !1,
  hasSelectedRows: [],
  sortFieldKey: "sortField",
  sortOrderKey: "sortOrder"
}, So = ["10", "20", "50"], Po = (t, e) => {
  var R, A;
  const i = Q(!1), n = Q(!1), r = Q([]), a = xt("speed-components-config"), s = (R = a == null ? void 0 : a.value) == null ? void 0 : R.transformRequestRes, u = (A = a == null ? void 0 : a.value) == null ? void 0 : A.useLoadConfig, o = Oe(() => ({ ..._o, ...e.value, ...u })), f = Oe(() => o.value.rowKey || "id"), c = Q({}), l = Q({
    [String(o.value.sortFieldKey)]: "",
    [String(o.value.sortOrderKey)]: ""
  }), d = o.value.emit, b = Q({
    current: 1,
    pageSize: 10,
    total: 0,
    showSizeChanger: !0,
    pageSizeOptions: So,
    showTotal: (p) => `共${p}条`
  }), h = Yi({
    curPageKeys: [],
    selectedRowKeys: [],
    selectedRows: [],
    totalRows: [],
    hasSelectedRows: []
  }), E = Oe(() => h.totalRows.map((p) => p[f.value]) || []), g = Oe(() => ({
    checkAll: h.hasSelectedRows.length > 0 && h.hasSelectedRows.length === Number(h.totalRows.length),
    indeterminate: h.hasSelectedRows.length > 0 && h.hasSelectedRows.length < h.totalRows.length
  })), m = () => {
    o.value && o.value.handleMultiRes ? h.curPageKeys = o.value.handleMultiRes(r.value) : h.curPageKeys = r.value.map(
      (p) => p[f.value]
    ), h.selectedRows = h.hasSelectedRows.filter((p) => h.curPageKeys.includes(p[f.value])) || [], h.selectedRowKeys = h.selectedRows.map((p) => p[f.value]) || [];
  };
  dn(
    () => o.value.hasSelectedRows,
    (p) => {
      h.hasSelectedRows = p || [], m();
    },
    { immediate: !0 }
  );
  const _ = async (p) => {
    p.target.checked ? h.hasSelectedRows = js(
      [...h.hasSelectedRows, ...h.totalRows],
      f.value
    ) : p.target.checked || (h.hasSelectedRows = h.hasSelectedRows.filter((I) => !E.value.includes(I[f.value]))), m(), d && d("update:hasSelectedRows", h.hasSelectedRows);
  };
  async function S() {
    if (o.value.fullRowsAjax) {
      n.value = !0;
      const p = await o.value.fullRowsAjax({
        ...o.value.extraParams
      });
      n.value = !1, p && p.success && (h.totalRows = p.data || [], m());
    }
  }
  const P = () => {
    h.curPageKeys = [], h.selectedRowKeys = [], h.selectedRows = [], h.totalRows = [], h.hasSelectedRows = [];
  };
  async function N(p = !0, I = {
    needResetChecked: !1,
    backLoad: !1
  }) {
    const k = Fi(t);
    i.value = !0, I.needResetChecked && P(), o.value.beforeFetch && typeof o.value.beforeFetch == "function" && o.value.beforeFetch(), p && (b.value.current = 1);
    try {
      const T = {
        ...o.value.extraParams,
        ...c.value,
        ...o.value.hasSort ? l.value : {}
      };
      let v = {};
      if (v = await k(
        o.value.hasPagination ? {
          page: b.value.current,
          size: b.value.pageSize,
          ...T
        } : T
      ), s && (typeof s != "function" ? console.error("transformRequestRes应为一个函数") : v = s(v)), v && (v != null && v.success)) {
        if (i.value = !1, r.value.length === 0 && b.value.current !== 1 && !I.backLoad) {
          b.value.current--, N(p, {
            backLoad: !0
          });
          return;
        }
        o.value.transformAfterFetch && typeof o.value.transformAfterFetch == "function" && (v.data = o.value.transformAfterFetch(v.data)), r.value = (v == null ? void 0 : v.data) ?? [], b.value.total = Number(v.totalCount), o.value.needFullSelect && p && S(), m(), o.value.afterFetch && typeof o.value.afterFetch == "function" && o.value.afterFetch(r.value);
      }
    } catch (T) {
      r.value = [], i.value = !1, console.error(T);
    }
  }
  return {
    state: h,
    totalCheckStatus: g,
    loading: i,
    selectAllLoading: n,
    pagination: b,
    dataSource: r,
    getList: N,
    onSelectChange: (p, I) => {
      console.log(p, I), h.selectedRowKeys = p, h.selectedRows = I;
      const k = h.hasSelectedRows.filter((T) => !h.curPageKeys.includes(T[f.value]));
      h.hasSelectedRows = [...k, ...I], d && d("update:hasSelectedRows", h.hasSelectedRows);
    },
    handleTableChange: (p, I, k) => {
      console.log(I, k), b.value.current = p.current, b.value.pageSize = p.pageSize, I && (c.value = I), k && (l.value[String(o.value.sortFieldKey)] = k.field, l.value[String(o.value.sortOrderKey)] = k.order), N(!1);
    },
    handleResizeColumn: (p, I) => {
      I.width = p;
    },
    onCheckAllChange: _
  };
}, Eo = {
  extraParams: {},
  hasPagination: !0
}, No = (t, e = Q({
  extraParams: {},
  hasPagination: !0
})) => {
  var E, g;
  const i = Fi(t), n = xt("speed-components-config"), r = (E = n == null ? void 0 : n.value) == null ? void 0 : E.transformRequestRes, a = (g = n == null ? void 0 : n.value) == null ? void 0 : g.useLoadConfig, s = Oe(() => ({ ...Eo, ...e.value, ...a })), u = Yi({
    page: 1,
    size: 10
  }), o = Q(!1), f = Q(!1), c = Q(!1), l = Q([]), d = async () => {
    o.value = !0;
    try {
      let m = await i({
        [s.value.pageKey]: u.page,
        [s.value.pageSizekey]: u.size,
        ...s.value.extraParams
      });
      if (r && (typeof r != "function" ? console.error("transformRequestRes应为一个函数") : m = r(m)), m && m.success) {
        o.value = !1, s.value.transformAfterFetch && typeof s.value.transformAfterFetch == "function" && (m.data = s.value.transformAfterFetch(m.data));
        const _ = l.value.concat(m.data);
        l.value = _, m.totalPages !== void 0 ? m.totalPages <= u.page ? c.value = !0 : c.value = !1 : m.totalCount !== void 0 && (m.totalCount <= l.value.length ? c.value = !0 : c.value = !1);
      } else
        o.value = !1;
    } catch (m) {
      o.value = !1, console.error(m);
    }
  };
  return {
    noMore: c,
    loading: o,
    initLoading: f,
    initLoad: async (m = !0) => {
      m && (u.page = 1, l.value = []), f.value = !0, await d(), f.value = !1;
    },
    onLoadMore: async () => {
      o.value = !0, c.value || (u.page += 1, d());
    },
    list: l
  };
}, { defaultAlgorithm: To, defaultSeed: Do } = vn, Ao = (t) => t.replace(/([a-z0-9])([A-Z])/g, "$1-$2").toLowerCase(), zo = () => {
  const t = To(Do), e = () => {
    const i = [];
    return Object.entries(t).forEach(([n, r]) => {
      if (typeof r == "string" || typeof r == "number") {
        const a = Ao(n);
        i.push(`--ant-${a}: ${r};`);
      }
    }), `:root {
  ${i.join(`
  `)}
}`;
  };
  if (typeof window < "u") {
    const i = document.createElement("style");
    return i.id = "antd-css-vars", i.textContent = e(), document.head.appendChild(i), () => {
      const n = document.getElementById("antd-css-vars");
      n && document.head.removeChild(n);
    };
  }
  return () => {
  };
};
export {
  mn as A,
  ye as S,
  ri as U,
  hn as V,
  Po as a,
  No as b,
  zo as c,
  So as d,
  Hs as e,
  $o as f,
  Lo as g,
  Us as h,
  Ct as i,
  It as j,
  Fe as k,
  Xi as l,
  ar as m,
  Re as n,
  Ur as o,
  we as p,
  Fr as q,
  de as r,
  Kt as s,
  _t as t,
  Ro as u,
  ce as v,
  Jt as w,
  fe as x,
  Io as y,
  Co as z
};
